---
title: "Dilution series analysis"
author: "Vincent Caruso"
date: "August 25, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up the environment.
```{r setup}

library(dada2)
library(stringr)
library(Biostrings)
library(tidyverse)
library(ShortRead)
library(RColorBrewer)
library(ggpubr)

options(tibble.print_min = 40, tibble.print_max = 40)

working_path <- "~/repo_test"
data_path <- file.path(working_path, "data", "dilution", "dilution_separate")
result_path <- file.path(working_path, "results", "dilution", "dilution_separate")
analysis_path <- file.path(working_path, "analysis","dilution", "dilution_separate")
ref_path <- file.path(working_path, "references")
combined_path <- file.path(result_path, "combined")
blast_path <- file.path(analysis_path, "blast")
summary_path <- file.path(analysis_path, "summary")
plot_path <- file.path(analysis_path, "plots")

script_path <- "~/thesis/noisy-microbes/community-inference/scripts"
source(file.path(script_path, "seq_table_processing.R"))

if (!file_test("-d", analysis_path)) dir.create(analysis_path, recursive = TRUE)
if (!file_test("-d", combined_path)) dir.create(combined_path)
if (!file_test("-d", blast_path)) dir.create(blast_path, recursive = TRUE)
if (!file_test("-d", summary_path)) dir.create(summary_path, recursive = TRUE)
if (!file_test("-d", plot_path)) dir.create(plot_path, recursive = TRUE)

```

##Load and format OTU/SV tables from each method

Load and inspect the tables from each clustering method. Modify tables as necessary so that they all have the following format:

1. Rows are OTUs or SVs
2. The first column is the OTU/SV identifier, and the remaining columns are samples' abundances
3. There are no comment lines (just a header with column names)

After modifying the format of each table, write it back to a tab-separated '.txt' file.
```{r load inference results}

methods <- c("uclust", "uparse","med", "unoise", "deblur", "dada2")
# methods <- c("uclust", "uparse", "mothur","med", "unoise", "deblur", "dada2")
samples <- str_subset(list.files(result_path), "^s\\d{3}")

for (m in methods){
  if (!file_test("-d", file.path(combined_path, m))) dir.create(file.path(combined_path, m))
}

```


```{r load UCLUST}

# load UCLUST table
uclust_otu_file <- "uclust/otu_table.txt"
uclust_seqs_file <- "uclust/rep_set/pooled_nochim_rep_set.fasta"
uclust_table <- list()

for (s in samples){
  uclust_table[[s]] <- load_uclust(file.path(result_path, s, uclust_otu_file), file.path(result_path, s, uclust_seqs_file))
}
names(uclust_table) <- str_replace_all(names(uclust_table), '-', '\\.')

uclust_table <- merge_tables(uclust_table, sample_names = names(uclust_table), collapse = FALSE, id = "keep")
sample_names <- colnames(uclust_table)[c(-1, -length(colnames(uclust_table)))]
uclust_table <- remove_rare(uclust_table, sample_names, min_abund = 2)  # remove singletons from all samples
uclust_table
write_tsv(uclust_table %>% select(-sequence), file.path(combined_path, "uclust/uclust_table.txt"))

```


```{r load UPARSE}

# load UPARSE table
uparse_otu_file <- "uparse/otu_table.txt"
uparse_seqs_file <- "uparse/otus.fa"
uparse_table <- list()

for (s in samples){
  uparse_table[[s]] <- load_uparse(file.path(result_path, s, uparse_otu_file),
                                   file.path(result_path, s, uparse_seqs_file))
}
names(uparse_table) <- str_replace_all(names(uparse_table), '-', '\\.')

uparse_table <- merge_tables(uparse_table, sample_names = names(uparse_table), collapse = FALSE, id = "keep")
uparse_table <- remove_rare(uparse_table, sample_names, min_abund = 2)  # remove singletons
uparse_table
write_tsv(uparse_table %>% select(-sequence), file.path(combined_path, "uparse/uparse_table.txt"))

```


```{r load MED}

# load MED table
med_otu_file <- "med/MATRIX-COUNT.txt"
med_seqs_file <- "med/NODE-REPRESENTATIVES.fasta"
med_chimera_file <- "med/uchime/NODE-chimeras.fasta"
med_table <- list()

for (s in samples){
  med_table[[s]] <- load_med(file.path(result_path, s, med_otu_file), 
                             file.path(result_path, s, med_seqs_file), 
                             file.path(result_path, s, med_chimera_file))
}
names(med_table) <- str_replace_all(names(med_table), "-", "\\.")

med_table <- merge_tables(med_table, sample_names = names(med_table), collapse = FALSE, id = "keep")
med_table <- remove_rare(med_table, sample_names, min_abund = 2)  # remove singletons
med_table
write_tsv(med_table %>% select(-sequence), file.path(combined_path, "med/med_table.txt"))

```


```{r load UNOISE}

# load UNOISE table
unoise_otu_file <- "unoise/zotu_table.txt"
unoise_seqs_file <- "unoise/zotus.fa"
unoise_table <- list()

for (s in samples){
  unoise_table[[s]] <- load_unoise(file.path(result_path, s, unoise_otu_file),
                                   file.path(result_path, s, unoise_seqs_file))
}
names(unoise_table) <- str_replace_all(names(unoise_table), "-", "\\.")

unoise_table <- merge_tables(unoise_table, sample_names = names(unoise_table), collapse = FALSE, id = "keep")
unoise_table <- remove_rare(unoise_table, sample_names, min_abund = 2)  # remove singletons
unoise_table
write_tsv(unoise_table %>% select(-sequence), file.path(combined_path, "unoise/unoise_table.txt"))

```


```{r load Deblur table}

# load Deblur table
deblur_otu_file <- "deblur/all.txt"
deblur_table <- list()

for (s in samples){
  deblur_table[[s]] <- load_deblur(file.path(result_path, s, deblur_otu_file))
}
names(deblur_table) <- str_replace_all(names(deblur_table), "-", "\\.")

deblur_table <- merge_tables(deblur_table, sample_names = names(deblur_table), collapse = FALSE, id = "keep")

deblur_table <- remove_rare(deblur_table, sample_names, min_abund = 2)  # remove singletons
deblur_table
write_tsv(deblur_table %>% select(-sequence), file.path(combined_path, "deblur/deblur_table.txt"))

```


```{r load DADA2 table}

# load DADA2 table
dada2_otu_file <- file.path(result_path, "dada2/sv_table.no_chim.txt")
dada2_table <- load_dada2(dada2_otu_file)
dada2_table <- remove_rare(dada2_table, sample_names, min_abund = 2)  # remove singletons
dada2_table
write_tsv(dada2_table %>% select(-sequence), file.path(combined_path, "dada2/dada2_table.txt"))

```


##Create and annotate a list of method tables

Create a list of method tables.
```{r create method table list}

# method_tables <- list(uclust_table, uparse_table, mothur_table, med_table, unoise_table, deblur_table, dada2_table)
method_tables <- list(uclust_table, uparse_table, med_table, unoise_table, deblur_table, dada2_table)
names(method_tables) <- methods

```


Read in reference sequences, and compute Levenshtein distances of inferred sequences to the reference sequences for each method.
```{r compute distance to references}

ref_fasta <- "Zymo_16S_references_v2.fasta"
ref_seqs <- readDNAStringSet(file.path(ref_path, ref_fasta)) %>% as.character()
names(ref_seqs) <- str_replace(names(ref_seqs), "_16S.*", "")

ref_dist_tables <- lapply(method_tables, compute_ref_dist, ref_seqs)
strain_dist_tables <- lapply(ref_dist_tables, collapse_group_dist, names(ref_seqs))


```


Annotate sequences in sequence tables as "reference" if they exactly match a reference sequence, or as "ref_noisy" if they are within a specified Levenshtein distance range from a reference sequence. 
```{r annotate reference}

max_dist = 10
method_tables <- mapply(annotate_ref, method_tables, ref_dist_tables, 
                        MoreArgs = list(sample_names = sample_names, max_dist = max_dist),
                        SIMPLIFY = FALSE)

```


Next, write each method's inferred (singleton-free) sequences to a fasta file, for BLASTing against nt.
```{r write fastas}

mapply(write_fasta, method_tables, file.path(blast_path, paste(methods, "seqs.fasta", sep = "_")))

# BLAST each method's sequences against the 'nt' database using the following command:
# blastn -query method_seqs.fasta -db nt -out method_blast.txt \
#        -task megablast -max_target_seqs 10 \
#        -outfmt "7 qseqid qlen sseqid slen sskingdoms ssciname pident length nident mismatch gapopen gaps qstart qend sstart send evalue bitscore" \

setwd(blast_path)
system2("blast_all.sh")

```


Read in the results of the BLAST searches.
```{r load blast tables}

blast_files <- file.path(blast_path, paste(methods, "blast.txt", sep = "_"))
# blast_files <- list.files(blast_path, ".+_blast.txt$")
blast_tables <- lapply(blast_files, load_blast)
names(blast_tables) <- methods
blast_tables <- lapply(blast_tables, annotate_blast_table)

```


Annotate sequences with distance to the nearest NT sequence, and label them as 'Contaminant' if they exactly match an NT sequence, using the results of the BLAST search.
```{r annotate contaminant}

method_tables <- mapply(annotate_nt_dist, method_tables, blast_tables, SIMPLIFY = FALSE)

method_tables <- mapply(annotate_contam, method_tables, blast_tables, 
                        MoreArgs = list(sample_names = sample_names, max_dist = 10), 
                        SIMPLIFY = FALSE)

```


Compute distance from sequences to Contaminant sequences.
```{r compute distance to contams}

contam_seqs <- lapply(method_tables, function(mt) return(mt %>% filter(contaminant) %>% .[["sequence"]]))
contam_dist_tables <- mapply(compute_ref_dist, method_tables, contam_seqs, SIMPLIFY = FALSE)

```


Annotate Contam_Noisy, Other, and Consensus, then add a 'Class' factor column with the sequence classification of each sequence, with the five main classification levels.
```{r annotate contam noisy}

method_tables <- mapply(annotate_contam_like, method_tables, contam_dist_tables, blast_tables,
                        MoreArgs = list(sample_names = sample_names, noisy_dist = 10, mut_dist = 1),
                        SIMPLIFY = FALSE)

method_tables <- lapply(method_tables, annotate_class)

```


##Create a list of tables that summarize method results

Compute summary tables for each method, with total sequence counts, sequence counts in each class, and percentages of reads in each class.
```{r compute method summaries}

method_summary <- mapply(summarize_seqs, method_tables, ref_dist_tables, 
                         MoreArgs = list(refs = ref_seqs, sample_names = sample_names, max_dist = max_dist),
                         SIMPLIFY = FALSE)
# names(method_summary) <- c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2")
sample_summary <- transpose_table_list(method_summary, "sample", "method")
sample_summary_short <- lapply(sample_summary, function(ss){
  return(ss %>% select(method, total, reference, ref_noisy, contaminant, contam_noisy, other))
})

method_summary <- lapply(method_summary, sanity_check_summary)
sample_summary <- lapply(sample_summary, sanity_check_summary)

lapply(method_summary, function(ms) sum(ms$total != ms$check_sum))
lapply(sample_summary, function(ss) sum(ss$total != ss$check_sum))

write_tables(method_summary, file.path(summary_path, "method_summary_tables.tsv"))
write_tables(sample_summary, file.path(summary_path, "sample_summary_tables.tsv"))
write_tables(sample_summary_short, file.path(summary_path, "sample_summary_short_tables.tsv"))

```


From the tables, compute recall and precision for each method, using both sequences and reads as the unit of analysis.
```{r method summary stats}

method_seq_stats <- lapply(method_summary, compute_pr_seqs)
sample_seq_stats <- lapply(sample_summary, compute_pr_seqs)

method_read_stats <- lapply(method_tables, compute_pr_reads, sample_names)
sample_read_stats <- transpose_table_list(method_read_stats, "sample", "method")

method_ref_perc <- lapply(method_tables, compute_ref_perc, sample_names = sample_names)
method_ref_perc <- do.call("rbind", method_ref_perc) %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()

write_tables(method_seq_stats, file.path(summary_path, "method_seq_stats.tsv"))
write_tables(sample_seq_stats, file.path(summary_path, "sample_seq_stats.tsv"))
write_tables(method_read_stats, file.path(summary_path, "method_read_stats.tsv"))
write_tables(sample_read_stats, file.path(summary_path, "sample_read_stats.tsv"))
write_tsv(method_ref_perc, file.path(summary_path, "method_ref_perc.tsv"))

```


##Create various plots to visualize results

Add a column that assigns a color code to each sequence
```{r add color column}

library(RColorBrewer)

ref_colors <- brewer.pal(8, "Dark2") %>% rev()
ref_colors <- c(ref_colors, brewer.pal(3, "Set1"))
max_contams <- max(sapply(method_tables, function(mt) nrow(mt[!mt$reference,])))
#other_colors <- sample(grey.colors(5, start = 0.5, end = 0.9), sum(!data_table$reference), replace = TRUE)
other_colors <- rep("#A1A1A1", times = max(sapply(method_tables, function(dt) nrow(dt[!dt$reference,]))))

method_tables <- lapply(method_tables, function(mt){
  mt$color <- character(length = nrow(mt))
  mt$color[mt$reference] <- ref_colors[1:nrow(mt[mt$reference,])]
  mt$color[!mt$reference] <- other_colors[1:nrow(mt[!mt$reference,])]
  return(mt)
})

```


Reformat annotated sequence tables for plotting with `ggplot`.
```{r prep for ggplot}

# Combine all method tables into one large table, then gather the sample columns into a single column
method_tables_gg <- lapply(method_tables, gather_samples, sample_names, sample)
method_tables_gg <- lapply(method_tables_gg, annotate_norms, group = "sample")
all_table_gg <- do.call("rbind", method_tables_gg) %>% mutate(method = rep(methods, times = sapply(method_tables_gg, nrow)))
all_table_gg$method <- factor(all_table_gg$method, levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2"),
                              labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2"))

# add some grouping variables for plotting convenience
all_table_gg <- all_table_gg %>% mutate(algorithm = method)
levels(all_table_gg$algorithm) <- list(OTU = c("UCLUST", "UPARSE"), entropy = "MED", EM = c("UNOISE", "Deblur", "DADA2"))

# Grab an arbitrary zero-count, non-Reference row from Zymo for Deblur, to use as a dummy filler
# This will be used for the SNR boxplot, since Deblur had no non-Reference sequences for Zymo
deblur_zero <- all_table_gg %>% filter(method == "Deblur", sample == "s160.MC.Neat", count == 0, class != "reference") %>% .[1, ]
deblur_one <- deblur_zero
deblur_one$count <- 1

# Remove sample rows for which the ASV/OTU count is zero
all_table_gg <- all_table_gg %>% filter(count > 0)

# Create plotting labels
dilution_labels <- sapply(str_split(sample_names, "\\."), function(d){
  paste(d[3:4], collapse = ":")
})
dilution_labels[1] <- "1:1"
dilution_values <- sapply(str_split(dilution_labels, ":"), function(d) as.numeric(d[1]) / as.numeric(d[2]))


```


Create tables of sequence and read count summaries per sample and method.
```{r summary tables}

load(file.path(result_path, "dada2/tables.RData"))
median_raw <- median(track_table[, "raw"])

sample_seq_count_gg <- sapply(method_tables, function(st, ss) colSums(st[, ss] > 0), sample_names) %>%
  t() %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()
sample_seq_count_gg <- gather(sample_seq_count_gg, "sample", "seqs", -method)

sample_read_count_gg <- sapply(method_tables, function(st, ss) colSums(st[, ss]), sample_names) %>%
  t() %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()
sample_read_count_gg <- gather(sample_read_count_gg, "sample", "reads", -method)

sample_counts_gg <- merge(sample_read_count_gg, sample_seq_count_gg, by = c("method", "sample"))
sample_counts_gg <- sample_counts_gg %>% group_by(method) %>%
  mutate(seqs_norm_method_median = round(seqs * median(reads) / reads),
         seqs_norm_raw_median = round(seqs * median_raw / reads)) %>%
  ungroup()

sample_counts_gg$method <- factor(sample_counts_gg$method, levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2"),
                              labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2"))
sample_counts_gg <- sample_counts_gg %>% mutate(algorithm = factor(method))
levels(sample_counts_gg$algorithm) <- list(OTU = c("UCLUST", "UPARSE"), entropy = "MED", EM = c("UNOISE", "Deblur", "DADA2"))

```


Plot composition of samples in terms of ASV/OTU classification, facetted by method.
```{r class sequences composition}

theme_set(theme_bw())
class_colors <- brewer.pal(12, "Paired")[c(2, 1, 8, 7, 10)]
classes = levels(all_table_gg$class)

(class_seqs_comp <- ggplot(all_table_gg, aes(x = sample)) +
  geom_bar(aes(fill = fct_rev(class)), width = 0.7) +
  facet_wrap(~method) +
  scale_fill_manual(name = "ASV/OTU class   ", values = rev(class_colors), labels = rev(classes)) +
  scale_x_discrete(labels = dilution_labels) +
  # labs(title = "Composition of dilution series communities by sequence class", subtitle = "", 
  labs(x = "", y = "number of ASVs/OTUs") +
  big_labels() +
  theme(plot.margin = margin(0, 2, 0, 2)) +
  guides(fill = guide_legend(reverse = TRUE, keywidth = 1, keyheight = 1)))

(class_reads_comp <- ggplot(all_table_gg, aes(x = sample)) +
  geom_col(aes(y = count, fill = fct_rev(class)), position = position_fill(), width = 0.7) +
  scale_x_discrete(labels = dilution_labels) +
  scale_fill_manual(name = "ASV/OTU class   ", values = rev(class_colors), labels = rev(classes)) +
  facet_wrap(~method, nrow = 2) +
  # labs(title = "Read-level composition by sequence class across dilution series", subtitle = "", 
  labs(x = "sample dilution", y = "fraction of sample") +
  big_labels() +
  theme(plot.margin = margin(6, 2, 0, 2)) +
  # guides(fill = guide_legend(reverse = TRUE, keywidth = 4.5, keyheight = 4.5))) +
  guides(fill = FALSE))

(class_comp_combined <- ggarrange(class_seqs_comp, class_reads_comp, labels = c("A", "B"), vjust = c(2.0, 0.5), nrow = 2, 
                                 font.label = list(size = 16), heights = c(5, 4.5)))

ggsave("dilution_compositions_combined.tif", plot = class_comp_combined, device = "tiff", path = plot_path, 
       width = 6.87, height = 8.0, units = "in", dpi = 300)

```


Plot numbers of Contaminant, Contam_Noisy, and Other ASVs/OTUs, faceted by method, to show whether these categories exhibit the same trend or not.
```{r non-reflike and reflike sequence lines}

ref_class_colors <- class_colors[1:2]
contam_colors <- class_colors[3:5]

(reflike_seq_lines <- dilution_line_plot(all_table_gg, "class", ref_like, "count", size = 0.7) +
  scale_color_manual(name = "ASV/OTU class   ", values = ref_class_colors, labels = paste(classes[1:2], "   ")) +
  labs(y = "number of ASVs/OTUs") +
  facet_wrap(~method) +
  big_labels() +
  theme(axis.title.x = element_blank(),
        plot.margin = margin(0, 2, 0, 2),
        legend.box.margin = margin(0, 0, -4, 0)) +
  guides(color = guide_legend(keywidth = 1.5)))

(non_reflike_seq_lines <- dilution_line_plot(all_table_gg, "class", !ref_like, "count", size = 0.7) +
  scale_color_manual(name = "ASV/OTU class   ", values = contam_colors, labels = paste(classes[3:5], "   ")) +
  labs(x = "sample dilution", y = "number of ASVs/OTUs") +
  facet_wrap(~method) +
  big_labels() +
  theme(legend.box.margin = margin(0, 0, -4, 0),
        plot.margin = margin(6, 2, 0, 2)) +
  guides(color = guide_legend(keywidth = 1.5)))

(all_seq_lines <- ggarrange(reflike_seq_lines, non_reflike_seq_lines, labels = c("A", "B"), 
                           vjust = c(1.7, 2.0), nrow = 2, font.label = list(size = 16)))

ggsave("all_seq_lines.tif", plot = all_seq_lines, device = "tiff", path = plot_path, 
       width = 6.87, height = 8.0, units = "in", dpi = 300)

```


Create plots to show relative strength of signal (reference sequence matches) to noise (non-reference sequences).
```{r snr plots}

library(scales)
library(ggnewscale)

(snr_box_facet <- ggplot(data = rbind(all_table_gg, deblur_one),
                        aes(x = sample, y = count)) +
  geom_boxplot(aes(fill = factor(reference, levels = c(TRUE, FALSE))),
               outlier.size = 0, #outlier.stroke = 1, size = 1.5
               size = 0.3, alpha = 1) +
  scale_fill_manual(name = "Sequence class", 
                    labels = c("Reference", "non-Reference"), 
                    values = c("blue", "orange")) +
  new_scale_fill() +
  geom_jitter(aes(fill = factor(reference, levels = c(TRUE, FALSE))),
              position = position_jitterdodge(jitter.width = 0.15, jitter.height = 0.), 
              shape = 21, size = 0.5, stroke = 0.25, color = "black", alpha = 1, show.legend = FALSE) +
  scale_fill_manual(values = c("deepskyblue", "gold1")) +
  facet_wrap(~factor(method), nrow = 3, dir = "v") +
  labs(x = "dilution", y = expression(bold("log"[10]*" abundance"))) +
  scale_x_discrete(labels = dilution_labels) +
  scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))) +
  big_labels() +
  # guides(fill = guide_legend(keywidth = 1, keyheight = 10)) +
  geom_line(aes(x=x, y=y),  # cover up the dummy value for Deblur
            data=data.frame(x=c(0.5, 1.5),y=rep(1,2),method="Deblur"),
            size = 1,
            col='white') +
  geom_line(aes(x=x, y=y),  # redraw the covered grid line
            data=data.frame(x=c(0.5, 1.5), y=rep(1, 2), method="Deblur"),
            size=0.5, alpha=0.4, col='grey'))

ggsave("dilution_snr_box_jitter.tif", plot = snr_box_facet, device = "tiff", path = plot_path, 
       width = 6.87, height = 8.0, units = "in", dpi = 300)

```


Create plot to show sample composition, colored by reference vs. non-reference.
```{r composition plot}

(comp_bar_facet <- ggplot(all_table_gg, aes(x = sample, y = count)) +
  geom_col(aes(fill = color, group = reference), width = 0.6, position = position_fill()) +
  scale_fill_identity(name = "ASV/OTU", guide = "legend",
                      labels = c(paste("Ref", 1:9), rep("non-Ref", length(unique(other_colors)))),
                      breaks = c(ref_colors[1:9], unique(other_colors))) +
  scale_x_discrete(labels = dilution_labels) +
  facet_wrap(~method) +
  # labs(title = "Relative abundance composition of dilution series samples", 
  labs(x = "sample dilution", y = "fraction of sample") +
  big_labels(legend.position = "right") +
  guides(fill = guide_legend(reverse = FALSE, keywidth = 1, keyheight = 1)))

```


Plot alpha diversity.
```{r alpha diversity}

library("phyloseq")

# extract sequence matrices of read counts from the method sequence tibbles
seq_mats <- lapply(method_tables, function(mt){
  seq_mat <- as.matrix(mt[, sample_names])
  rownames(seq_mat) <- mt$id
  return(seq_mat)
})


# construct phyloseq OTU and taxonomy tables
otu_tabs <- lapply(seq_mats, otu_table, taxa_are_rows = TRUE)

n_strains <- length(unique(names(ref_seqs)))
ref_props <- tibble(expected = 10000 * rep(1 / n_strains, times = n_strains))

# construct phyloseq objects
physeqs <- lapply(otu_tabs, phyloseq)
phy_exp <- phyloseq(otu_table(ref_props, taxa_are_rows = TRUE))

# compute alpha diversity
alpha_indices <- c("Shannon", "InvSimpson", "Fisher")
alpha_div <- lapply(physeqs, estimate_richness, measures = alpha_indices)
alpha_exp <- estimate_richness(phy_exp, measures = alpha_indices)

# combine alpha diversity tables into one table for plotting
alpha_div <- lapply(alpha_div, rownames_to_column, var = "sample")
alpha_div <- lapply(alpha_div, as.tibble)
alpha_div_gg <- lapply(alpha_div, gather, key = "alpha_index", value = "diversity", alpha_indices)
alpha_div_gg <- do.call("rbind", alpha_div_gg) %>% 
  mutate(method = rep(methods, each = length(sample_names) * length(alpha_indices)))

alpha_exp <- rownames_to_column(alpha_exp, var = "sample") %>% as.tibble()
alpha_exp_gg <- gather(alpha_exp, key = "alpha_index", value = "diversity", alpha_indices) %>%
  mutate(method = "expected") 

alpha_div_gg <- rbind(alpha_div_gg, alpha_exp_gg)


alpha_div_gg$method <- factor(alpha_div_gg$method, levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2", "expected"),
                              labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2", "Expected"))
alpha_div_gg$alpha_index <- factor(alpha_div_gg$alpha_index, levels = alpha_indices,
                                   labels = c("Shannon", "Inverse Simpson", "Fisher"))

method_colors <- brewer.pal(12, "Paired")[c(2, 3, 9, 7, 6, 12)]

(alpha_dots_sample <- ggplot(alpha_div_gg %>% filter(method != "Expected", sample != "expected"),
                     aes(x = sample)) +
  geom_point(aes(y = diversity, color = method), size = 3) +
  geom_hline(data = alpha_div_gg %>% filter(sample == "expected"), 
             aes(yintercept = diversity), color = "black", show.legend = TRUE) +
  geom_line(aes(y = diversity, color = method, group = method), size = 0.5) +
  scale_color_manual(name = "Method  ", values = method_colors, labels = levels(alpha_div_gg$method)) +
  scale_x_discrete(labels = dilution_labels) +
  labs(y = "alpha diversity") +
  facet_wrap(~alpha_index, nrow = 1, scales = "free_y") +
  big_labels() +
  guides(color = guide_legend(keywidth = 1.5, keyheight = 1, nrow = 1)))

ggsave("dilution_alpha_sample_x.tif", plot = alpha_dots_sample, device = "tiff", path = plot_path, 
       width = 6.87, height = 5.4, units = "in", dpi = 300)


sample_colors <- rev(brewer.pal(9, "YlOrRd"))

(alpha_dots_method <- ggplot(alpha_div_gg %>% filter(method != "Expected", sample != "expected"),
                     aes(x = method)) +
  geom_point(aes(y = diversity, color = sample), size = 3) +
  geom_hline(data = alpha_div_gg %>% filter(sample == "expected"), 
             aes(yintercept = diversity), color = "black") +
  scale_color_manual(name = "Sample  ", values = sample_colors, labels = paste(dilution_labels, "   ")) +
  scale_x_discrete(labels = levels(alpha_div_gg$method)) +
  labs(y = "alpha diversity") +
  facet_wrap(~alpha_index, nrow = 1, scales = "free_y") +
  big_labels() +
  theme(axis.title.y = element_text(vjust = -1),
        plot.margin = unit(c(2, 0, 2, 0), "lines")) +
  guides(color = guide_legend(nrow = 1)))

```


Plot number of inferred sequences for each sample and each method.
```{r sequence count plots}

# normalized by median read count after processing by each method
(seqs_line_norm <- ggplot(sample_counts_gg, aes(x = sample, y = seqs_norm_method_median)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method), size = 0.7) +
  # labs(title = "Number of inferred sequences at each concentration by each method", 
  #      subtitle = "Normalized by median sample read count",
  labs(x = "sample dilution", y = "normalized number of ASVs/OTUs") +
  scale_color_brewer(name = "Method   ", labels = levels(sample_counts_gg$method), palette = "Dark2") +
  scale_x_discrete(labels = dilution_labels) +
  big_labels() +
  theme(legend.title = element_text(margin = margin(r = 2))) +
  guides(color = guide_legend(nrow = 1, keywidth = 1.5)))

ggsave("dilution_seq_counts_norm.tif", plot = seqs_line_norm, device = "tiff", path = plot_path, 
       width = 6.87, height = 5.4, units = "in", dpi = 300)

```
