---
title: "High biomass analysis"
author: "Vincent Caruso"
date: "March 1, 2018"
output: html_document
---

Set up the environment, and define key variables.
```{r setup}

knitr::opts_chunk$set(echo = TRUE)

library(dada2)
library(stringr)
library(Biostrings)
library(tidyverse)
library(ShortRead)
library(RColorBrewer)
library(ggpubr)

options(tibble.print_min = 100, tibble.print_max = 40)

working_path <- "~/repo_test"
data_path <- file.path(working_path, "data", "high_biomass")
result_path <- file.path(working_path, "results", "high_biomass")
analysis_path <- file.path(working_path, "analysis", "high_biomass")
ref_path <- file.path(working_path, "references")
blast_path <- file.path(analysis_path, "blast")
summary_path <- file.path(analysis_path, "summary")
plot_path <- file.path(analysis_path, "plots")

script_path <- "~/thesis/noisy-microbes/community-inference/scripts"
source(file.path(script_path, "seq_table_processing.R"))

# if (!file_test("-d", combined_path)) dir.create(combined_path)
if (!file_test("-d", analysis_path)) dir.create(analysis_path)
if (!file_test("-d", blast_path)) dir.create(blast_path, recursive = TRUE)
if (!file_test("-d", summary_path)) dir.create(summary_path, recursive = TRUE)
if (!file_test("-d", plot_path)) dir.create(plot_path, recursive = TRUE)

datasets <- c("zymo", "kozich", "schirmer", "damore")
methods <- c("uclust", "uparse", "med", "unoise", "deblur", "dada2")

```


##Load OTU/SV tables from each method

Load and inspect the tables from each clustering method. Modify tables as necessary so that they all have the following format:

1. Rows are OTUs or SVs
2. The first column is the OTU/SV identifier, and the remaining columns are samples' abundances
3. There are no comment lines (just a header with column names)

After modifying the format of each table, write it back to a tab-separated '.txt' file.
```{r load inference results}

dataset_tables <- list()

for (d in datasets){
  resdir <- file.path(result_path, d)
  method_tables <- list()
  
  # load UCLUST table
  uclust_otu_file <- file.path(resdir, "uclust/otu_table.txt")
  uclust_seq_file <- file.path(resdir, "uclust/rep_set/pooled_nochim_rep_set.fasta")
  method_tables$uclust <- load_uclust(uclust_otu_file, uclust_seq_file)
  sample_name <- colnames(method_tables$uclust)[c(-1, -length(colnames(method_tables$uclust)))]
  method_tables$uclust <- remove_rare(method_tables$uclust, sample_name, min_abund = 2)  # remove singletons from all samples
  method_tables$uclust
  write_tsv(method_tables$uclust %>% select(-sequence), file.path(resdir, "uclust/uclust_table.txt"))
  
  # load UPARSE table
  uparse_otu_file <- file.path(resdir, "uparse/otu_table.txt")
  uparse_seq_file <- file.path(resdir, "uparse/otus.fa")
  method_tables$uparse <- load_uparse(uparse_otu_file, uparse_seq_file, sample_names = sample_name)
  method_tables$uparse <- remove_rare(method_tables$uparse, sample_name, min_abund = 2)  # remove singletons
  method_tables$uparse
  write_tsv(method_tables$uparse %>% select(-sequence), file.path(resdir, "uparse/uparse_table.txt"))
  
  # load MED table
  med_otu_file <- file.path(resdir, "med/MATRIX-COUNT.txt")
  med_seq_file <- file.path(resdir, "med/NODE-REPRESENTATIVES.fasta")
  med_chimera_file <- file.path(resdir, "med/uchime/NODE-chimeras.fasta")
  method_tables$med <- load_med(med_otu_file, med_seq_file, med_chimera_file)
  method_tables$med <- remove_rare(method_tables$med, sample_name, min_abund = 2)  # remove singletons
  method_tables$med
  write_tsv(method_tables$med %>% select(-sequence), file.path(resdir, "med/med_table.txt"))

  # load UNOISE table
  unoise_otu_file <- file.path(resdir, "unoise/zotu_table.txt")
  unoise_seq_file <- file.path(resdir, "unoise/zotus.fa")
  method_tables$unoise <- load_unoise(unoise_otu_file, unoise_seq_file, sample_names = sample_name)
  method_tables$unoise <- remove_rare(method_tables$unoise, sample_name, min_abund = 2)  # remove singletons
  method_tables$unoise
  write_tsv(method_tables$unoise %>% select(-sequence), file.path(resdir, "unoise/unoise_table.txt"))
  
  # load Deblur table
  deblur_otu_file <- file.path(resdir, "deblur/all.txt")
  method_tables$deblur <- load_deblur(deblur_otu_file)
  method_tables$deblur <- remove_rare(method_tables$deblur, sample_name, min_abund = 2)  # remove singletons
  method_tables$deblur
  write_tsv(method_tables$deblur %>% select(-sequence), file.path(resdir, "deblur/deblur_table.txt"))
  
  # load DADA2 table
  dada2_otu_file <- file.path(resdir, "dada2/sv_table.no_chim.txt")
  method_tables$dada2 <- load_dada2(dada2_otu_file, sample_names = sample_name)
  method_tables$dada2 <- remove_rare(method_tables$dada2, sample_name, min_abund = 2)  # remove singletons
  method_tables$dada2
  write_tsv(method_tables$dada2 %>% select(-sequence), file.path(resdir, "dada2/dada2_table.txt"))

  all_table <- merge_tables(method_tables, sample_name, collapse = TRUE)
  dataset_tables[[d]] <- all_table
}

```


Read in reference sequences, and compute Levenshtein ("Hamming") distances between all inferred sequences and all reference sequences.
```{r load reference fasta files}

zymo_fastas <- "Zymo_16S_references_v2.fasta"
kozich_fasta <- "kozich_HMP_MOCK.fasta"
schirmer_fasta <- "schirmer_BalancedRefSeqs.fasta"
damore_fasta <- "schirmer_BalancedRefSeqs.fasta"
ref_fastas <- list("zymo" = zymo_fastas, "kozich" = kozich_fasta, "schirmer" = schirmer_fasta, "damore" = damore_fasta)

ref_seqs <- lapply(ref_fastas, function(f){
  fp <- file.path(ref_path, f)
  seqs <- as.character(readDNAStringSet(fp))
  return(seqs)
})

names(ref_seqs$zymo) <- str_replace(names(ref_seqs$zymo), "_16S.*", "")
names(ref_seqs$kozich) <- str_replace(names(ref_seqs$kozich), "\\.\\d+$", "")

ref_dist_tables <- mapply(compute_ref_dist, dataset_tables, ref_seqs, SIMPLIFY = FALSE)
strain_dist_tables <- mapply(collapse_group_dist, ref_dist_tables, lapply(ref_seqs, names), SIMPLIFY = FALSE)

```


Check to see if any reference strains are completely absent from the dataset.
```{r missing strains}

# Start with just those reference strains not detected by any method
undetected <- lapply(strain_dist_tables, function(st){
  missing <- apply(st, 2, min) > 0
  missing <- names(missing[missing])
  return(missing)
})

# Get the corresponding reference sequences
undetected_refs <- mapply(function(refs, und){
  refs <- refs[names(refs) %in% und]
}, ref_seqs, undetected)

# Read in the (cleaned) fastq data
data_fastqs <- sapply(datasets, function(d) {
  fp <- file.path(data_path, d, "filtered")
  fastq <- list.files(fp, pattern = "\\.fastq$")
  return(file.path(fp, fastq))
})

data_seqs <- lapply(data_fastqs, function(fq){
  seqs <- derepFastq(fq, verbose = TRUE)
  seqs <- names(seqs$uniques)
  return(seqs)
})

missing_strains <- mapply(get_missing_strains, data_seqs, undetected_refs)

# Remove missing references from the reference set and ref_dist_tables
ref_seqs <- mapply(function(refs, missing){
  return(refs[!names(refs) %in% missing])
}, ref_seqs, missing_strains)

ref_dist_tables <- mapply(function(rdist, missing){
  return(rdist[, !colnames(rdist) %in% missing])
}, ref_dist_tables, missing_strains)

```


Annotate sequences in sequence tables as "Reference" if they exactly match a reference sequence, or as "Ref Noisy" if they are within a specified Levenshtein distance range from a reference sequence. 
```{r annotate reference and ref_noisy}

max_dist <- 10
dataset_tables <- mapply(annotate_ref, dataset_tables, ref_dist_tables, 
                         MoreArgs = list(sample_names = methods, max_dist = max_dist),
                         SIMPLIFY = FALSE)

```


Next, write each method's inferred (singleton-free) sequences to a fasta file, for BLASTing against nt.
```{r write fastas}

mapply(write_fasta, dataset_tables, file.path(blast_path, paste(datasets, "seqs.fasta", sep = "_")))

# BLAST sequences against the 'nt' database using the following command:
# blastn -query all_seqs.fasta -db nt -out all_blast.txt \
#        -task megablast -max_target_seqs 10 \
#        -outfmt "7 qseqid qlen sseqid slen sskingdoms ssciname pident length nident mismatch gapopen gaps qstart qend sstart send evalue bitscore" \
#        -remote

setwd(blast_path)
system2("blast_all.sh")

```


Read in the results of the BLAST searches. 
```{r load and annotate blast tables}

blast_files <- file.path(blast_path, paste(datasets, "blast.txt", sep = "_"))
blast_tables <- lapply(blast_files, load_blast)
names(blast_tables) <- datasets
blast_tables <- lapply(blast_tables, annotate_blast_table)

```


Annotate sequences with distance to the nearest NT sequence, and label them as 'Contaminant' if they exactly match an NT sequence, using the results of the BLAST search.
```{r annotate contaminant}

dataset_tables <- mapply(annotate_nt_dist, dataset_tables, blast_tables, SIMPLIFY = FALSE)

dataset_tables <- mapply(annotate_contam, dataset_tables, blast_tables, 
                        MoreArgs = list(sample_names = methods, max_dist = 10), 
                        SIMPLIFY = FALSE)

```


Compute distance from sequences to Contaminant sequences.
```{r compute distance to contaminant}

contam_seqs <- lapply(dataset_tables, function(dt) return(dt %>% filter(contaminant) %>% .[["sequence"]]))
contam_dist_tables <- mapply(compute_ref_dist, dataset_tables, contam_seqs, SIMPLIFY = FALSE)

```


Annotate Contam_Noisy, Other, and Consensus, then add a 'Class' factor column with the sequence classification of each sequence, with the five main classification levels.
```{r annotate contam noisy and contam mutant}

dataset_tables <- mapply(annotate_contam_like, dataset_tables, contam_dist_tables, blast_tables,
                        MoreArgs = list(sample_names = methods, noisy_dist = 10, mut_dist = 1),
                        SIMPLIFY = FALSE)

dataset_tables <- lapply(dataset_tables, annotate_class)

```


Compute a table of sequence summaries for each method.
```{r compute sequence summaries}


dataset_summary <- mapply(summarize_seqs, dataset_tables, ref_dist_tables, ref_seqs,
                              MoreArgs = list(sample_names = methods, max_dist = max_dist),
                              SIMPLIFY = FALSE)
dataset_summary <- lapply(dataset_summary, function(ds){
  colnames(ds)[1] <- "method"
  return(ds)
})
dataset_summary <- lapply(dataset_summary, sanity_check_summary)
lapply(dataset_summary, function(ds) sum(ds$total != ds$check_sum))


dataset_summary_short <- lapply(dataset_summary, function(ds){
  dss <- ds %>% select(method, total, reference, ref_noisy, contaminant, contam_noisy, other)
  dss$method <- c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2")
  return(dss)
})

write_tables(dataset_summary, file.path(summary_path, "dataset_summary_tables.tsv"))
write_tables(dataset_summary_short, file.path(summary_path, "dataset_summary_short_tables.tsv"), compact = FALSE)


```


From the summary table, compute recall and precision for the sequences inferred by each method.
```{r method sequence summary stats}

dataset_seq_stats <- lapply(dataset_summary, compute_pr_seqs)
dataset_read_stats <- lapply(dataset_tables, compute_pr_reads, methods)
dataset_ref_perc <- lapply(dataset_tables, compute_ref_perc, sample_names = methods)
dataset_ref_perc <- do.call("cbind", dataset_ref_perc) %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as_tibble()

dataset_seq_stats <- lapply(dataset_seq_stats, function(dss){
  dss$method <- c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2")
  return(dss)
})

write_tables(dataset_seq_stats, file.path(summary_path, "dataset_seq_stats.tsv"))
write_tables(dataset_read_stats, file.path(summary_path, "dataset_read_stats.tsv"))
write_tsv(dataset_ref_perc, file.path(summary_path, "dataset_ref_perc.tsv"))


```


Add a column that assigns a color code to each sequence
```{r add color column}

library(RColorBrewer)

ref_colors <- brewer.pal(8, "Dark2") %>% rev()
ref_colors <- c(brewer.pal(8, "Dark2") %>% rev(), brewer.pal(9, "Set1"), brewer.pal(8, "Accent") %>% rev(),
                brewer.pal(8, "Set2"), brewer.pal(12, "Paired"), brewer.pal(12, "Set3"), brewer.pal(11, "Spectral"))
ref_colors <- c(ref_colors, brewer.pal(3, "Set1"))
max_contams <- max(sapply(dataset_tables, function(mt) nrow(mt[!mt$reference,])))
other_colors <- sample(grey.colors(5, start = 0.5, end = 0.9), size = max_contams, replace = TRUE)
other_colors <- rep("#000000", times = max(sapply(dataset_tables, function(dt) nrow(dt[!dt$reference,]))))

dataset_tables <- lapply(dataset_tables, function(dt){
  dt$color <- character(length = nrow(dt))
  dt$color[dt$reference] <- ref_colors[1:nrow(dt[dt$reference,])]
  dt$color[!dt$reference] <- other_colors[1:nrow(dt[!dt$reference,])]
  return(dt)
})

```


Reformat annotated sequence tables for plotting with `ggplot`.
```{r prep for ggplot}

# Combine all method tables into one large table, then gather the sample columns into a single column
dataset_tables_gg <- lapply(dataset_tables, gather_samples, methods, method)
all_table_gg <- do.call("rbind", dataset_tables_gg) %>% mutate(dataset = rep(datasets, times = sapply(dataset_tables_gg, nrow)))
all_table_gg$dataset <- factor(all_table_gg$dataset, levels = c("zymo", "kozich", "schirmer", "damore"),
                              labels = c("Zymo", "Kozich", "Schirmer", "D'Amore"))
all_table_gg$method <- factor(all_table_gg$method, levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2"),
                              labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2"))

all_table_gg <- annotate_norms(all_table_gg, group = c("dataset", "method"))

# add some grouping variables for plotting convenience
all_table_gg <- all_table_gg %>% mutate(algorithm = method)
levels(all_table_gg$algorithm) <- list(OTU = c("UCLUST", "UPARSE"), entropy = "MED", EM = c("UNOISE", "Deblur", "DADA2"))

# Grab an arbitrary zero-count, non-Reference row from Zymo for Deblur, to use as a dummy filler
# This will be used for the SNR boxplot, since Deblur had no non-Reference sequences for Zymo
deblur_zero <- all_table_gg %>% filter(method == "Deblur", dataset == "Zymo", count == 0, class != "reference") %>% .[1, ]
deblur_zero$count <- 1

# Remove zero-count rows
all_table_gg <- all_table_gg %>% filter(count > 0)

```


##Create various plots to visualize results

Plot the sequence-level composition of samples by sequence class.
```{r class seqs composition}

theme_set(theme_bw())
classes = levels(all_table_gg$class)
class_colors <- brewer.pal(12, "Paired")[c(2, 1, 8, 7, 10)]

class_seqs_comp <- ggplot(all_table_gg, aes(x = method)) +
  geom_bar(aes(fill = fct_rev(class)), width = 0.65) +
  facet_wrap(~dataset, scales = "free_y") +
  scale_fill_manual(name = "ASV/OTU class  ", values = rev(class_colors), labels = rev(classes)) +
  labs(x = "method", y = "number of ASVs/OTUs") +
  big_labels() +
  theme(legend.box.margin = margin(b = 10)) +
  guides(fill = guide_legend(reverse = TRUE))
class_seqs_comp

ggsave("high_seq_classes.tif", plot = class_seqs_comp, device = "tiff", path = plot_path, 
       width = 6.87, height = 5.4, units = "in", dpi = 300)

class_reads_comp <- ggplot(all_table_gg, aes(x = method)) +
  geom_col(aes(y = count, fill = fct_rev(class)), position = position_fill(), width = 0.65) +
  scale_fill_manual(name = "ASV/OTU class  ", values = rev(class_colors), 
                    labels = rev(classes),
                    guide = guide_legend(reverse = FALSE)) +
  facet_wrap(~dataset) +
  # labs(title = "Read-level composition by sequence class",
  labs(x = "method", y = "fraction of sample") +
  big_labels() +
  theme(legend.box.margin = margin(b = 10)) +
  guides(fill = guide_legend(reverse = TRUE))
  # guides(fill = FALSE)
class_reads_comp

ggsave("high_read_composition.tif", plot = class_reads_comp, device = "tiff", path = plot_path, 
       width = 6.87, height = 5.4, units = "in", dpi = 300)

ggarrange(class_seqs_comp, class_reads_comp, labels = c("A", "B"), vjust = c(2.0, 0.5), nrow = 2, 
          font.label = list(size = 16), heights = c(5, 4.5))

```


Create plot to show sample composition, colored by reference vs. non-reference.
```{r composition plot}

(comp_bar_facet <- ggplot(all_table_gg, aes(x = method, y = count)) +
  geom_col(aes(fill = color, group = reference), width = 0.5, position = position_fill()) +
  scale_fill_identity(name = "ASV/OTU", guide = "legend",
                      labels = c(paste("Ref", 1:58), rep("non-Ref", length(other_colors))),
                      breaks = c(ref_colors[1:58], other_colors)) +
  scale_x_discrete(labels = methods) +
  facet_wrap(~dataset) +
  # labs(title = "Composition of high biomass samples", 
  labs(x = "method", y = "fraction of sample") +
  big_labels(legend.position = "right"))

```


Calculate expected proportions of each strain in each dataset sample, for use in expected alpha diversity calculations.
```{r reference proportions}

damore_distrib_file <- file.path(ref_path, "damore_proportions.tsv")

ref_props <- lapply(ref_seqs, function(rs){
  n_strains <- length(unique(names(rs)))
  props <- 10000 * rep(1 / n_strains, times = n_strains)
})

damore_props <- read_tsv(damore_distrib_file, col_names = c("name", "proportion"))
damore_props$name <- sapply(str_split(damore_props$name, " +"), paste, collapse = "_")
damore_props <- damore_props[damore_props$name %in% names(ref_seqs$damore), ]
damore_props$proportion <- 100 * damore_props$proportion

ref_props$damore <- damore_props$proportion

# construct table of expected strain proportions
max_refs <- max(sapply(ref_props, length))
ref_props <- lapply(ref_props, function(rp){
  rp <- as.integer(rp)
  length(rp) <- max_refs
  rp[is.na(rp)] <- 0
  return(rp)
  })
ref_props <- do.call("cbind", ref_props) 

```


Plot alpha diversity.
```{r alpha diversity}

library("phyloseq")

# extract sequence matrices of read counts only from the method sequence tibbles
seq_mats <- lapply(dataset_tables, function(mt){
  seq_mat <- as.matrix(mt[, methods])
  rownames(seq_mat) <- mt$id
  return(seq_mat)
})


# construct phyloseq OTU and taxonomy tables
otu_tabs <- lapply(seq_mats, otu_table, taxa_are_rows = TRUE)

# construct phyloseq objects
physeqs <- lapply(otu_tabs, phyloseq)
phy_exp <- phyloseq(otu_table(ref_props, taxa_are_rows = TRUE))

# compute alpha diversity
alpha_indices <- c("Shannon", "InvSimpson", "Fisher")
alpha_div <- lapply(physeqs, estimate_richness, measures = alpha_indices)
alpha_exp <- estimate_richness(phy_exp, measures = alpha_indices)

# combine alpha diversity tables into one table for plotting
alpha_div <- lapply(alpha_div, rownames_to_column, var = "method")
alpha_div <- lapply(alpha_div, as.tibble)
alpha_div_gg <- lapply(alpha_div, gather, key = "alpha_index", value = "diversity", alpha_indices)
alpha_div_gg <- do.call("rbind", alpha_div_gg) %>% 
  mutate(dataset = rep(datasets, each = length(methods) * length(alpha_indices)))

alpha_exp <- rownames_to_column(alpha_exp, var = "dataset") %>% as.tibble()
alpha_exp_gg <- gather(alpha_exp, key = "alpha_index", value = "diversity", alpha_indices) %>%
  mutate(method = "expected") %>% select(method, alpha_index, diversity, dataset)

alpha_div_gg <- rbind(alpha_div_gg, alpha_exp_gg)

alpha_div_gg$dataset <- factor(alpha_div_gg$dataset, levels = c("zymo", "kozich", "schirmer", "damore"),
                               labels = c("Zymo", "Kozich", "Schirmer", "D'Amore"))
alpha_div_gg$method <- factor(alpha_div_gg$method, levels = c("uclust", "uparse", "med", "unoise", "deblur", "dada2", "expected"),
                              labels = c("UCLUST", "UPARSE", "MED", "UNOISE", "Deblur", "DADA2", "Expected"))
alpha_div_gg$alpha_index <- factor(alpha_div_gg$alpha_index, levels = alpha_indices,
                                   labels = c("Shannon", "Inverse Simpson", "Fisher"))
alpha_div_gg <- alpha_div_gg %>% mutate(expected = method == "Expected")


alpha_colors <- brewer.pal(12, "Paired")[c(2, 4, 9, 8)]

(alpha_dots <- ggplot(alpha_div_gg %>% filter(alpha_index %in% c("Shannon", "Inverse Simpson", "Fisher")),
                     aes(x = method, shape = expected)) +
  geom_point(aes(y = diversity, fill = dataset, color = expected), size = 3, stroke = 1) +
  labs(y = "alpha diversity") +
  scale_fill_manual(name = "Dataset  ", values = alpha_colors, labels = paste(levels(alpha_div_gg$dataset), "  ")) +
  scale_shape_manual(name = " Source  ", values = c(21, 24), labels = paste(c("observed", "expected"), "  ")) +
  scale_color_manual(values = c("white", "black"), guide = FALSE) +
  facet_wrap(~alpha_index, nrow = 1, scales = "free_y") +
  big_labels() +
  guides(fill = guide_legend(override.aes = list(shape = 21, color = "white"))) +
  theme(axis.title.y = element_text(vjust = 1),
        legend.box = "vertical",
        legend.box.margin = margin(b = 10)))

ggsave("high_alpha.tif", plot = alpha_dots, device = "tiff", path = plot_path, 
       width = 6.87, height = 5.4, units = "in", dpi = 300)

ggsave("high_alpha.png", device = "png", path = plot_path, width = 7.5, height = 5.9, units = "in", dpi = 600)

```


Create plots to show relative strength of signal (reference sequence matches) to noise (non-reference sequences).
```{r snr plots}

library(scales)
library(ggnewscale)

(snr_box_facet <- ggplot(data = rbind(all_table_gg, deblur_zero),
                        aes(x = method, y = count)) +
  geom_boxplot(aes(fill = factor(reference, levels = c(TRUE, FALSE))),
               outlier.size = 0., outlier.color = "white", 
               size = 0.3, alpha = 1) +
  scale_fill_manual(name = "ASV/OTU class   ", 
                    labels = c(" Reference", " non-Reference"), 
                    values = c("blue", "orange")) +
  new_scale_fill() +
  geom_jitter(aes(fill = factor(reference, levels = c(TRUE, FALSE))),
              position = position_jitterdodge(jitter.width = 0.15, jitter.height = 0.),
              shape = 21, size = 0.5, stroke = 0.25, color = "black", alpha = 1, show.legend = FALSE) +
  scale_fill_manual(values = c("deepskyblue", "gold1")) +
  facet_wrap(~dataset) +
  labs(x = "method", y = expression(bold("log"[10]*" abundance"))) +
  scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x),
        labels = trans_format("log10", math_format(10^.x))
        ) +
  big_labels() +
  theme(legend.box.margin = margin(b = 10)) +
  geom_line(aes(x=x, y=y), 
            data=data.frame(x=c(4.5, 5.5),y=rep(1,2),dataset="Zymo"), 
            size = 1, 
            col='white') +
  geom_line(aes(x=x, y=y),
            data=data.frame(x=c(4.5, 5.5), y=rep(1, 2), dataset="Zymo"),
            size=0.5, alpha=0.4, col='grey'))

ggsave("high_snr_box_jitter.tif", plot = snr_box_facet, device = "tiff", path = plot_path, 
       width = 6.87, height = 5.4, units = "in", dpi = 300)

```
