---
title: "Dilution series analysis"
author: "Vincent Caruso"
date: "August 25, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Set up the environment.
```{r setup}

library(dada2)
library(stringr)
library(Biostrings)
library(tidyverse)
library(ShortRead)

result_path <- "~/thesis/results/dilution"
ref_path <- "~/thesis/references"
analysis_path <- "~/thesis/analysis/dilution"

if (!file_test("-d", analysis_path)) dir.create(analysis_path)

```

##Load OTU/SV tables from each method

Load and inspect the tables from each clustering method. Modify tables as necessary so that they all have the following format:

1. Rows are OTUs or SVs, and columns are samples
2. Row names are the OTU/SV identifiers
3. Column names are the sample names
4. There are no comment lines (just a header with column names)

After modifying the format of each table, write it back to a tab-separated '.txt' file.
```{r load UCLUST}

# load UCLUST table
uclust_table <- as.tibble(read.table(file = file.path(result_path, "uclust/otu_table.txt"), header = TRUE, sep = "\t", 
                                     skip = 1, comment.char = ""))
colnames(uclust_table)[1] <- "id"
uclust_table$id <- str_replace(uclust_table$id, "denovo", "denovo_")
sample_names <- colnames(uclust_table)[-1]
sample_names <- sample_names[order(sample_names)]
uclust_table <- uclust_table %>% select(id, noquote(sample_names))
uclust_table
write.table(uclust_table, file = file.path(result_path, "uclust/uclust_table.txt"), quote = FALSE, sep = "\t")

# load UCLUST representative sequences
uclust_seqs <- readDNAStringSet(file.path(result_path, "uclust/rep_set/pooled_nochim_rep_set.fasta"))
uclust_seqs <- as.character(uclust_seqs)
names(uclust_seqs) <- sapply(str_split(names(uclust_seqs), " "), `[`, 1)
names(uclust_seqs) <- str_replace(names(uclust_seqs), "denovo", "denovo_")

```


```{r load UPARSE}

# load UPARSE table
uparse_table <- as.tibble(read.table(file = file.path(result_path, "uparse/otu_table.txt"), header = TRUE, sep = "\t", 
                                     comment.char = ""))
colnames(uparse_table)[1] <- "id"
uparse_table$id <- str_replace(uparse_table$id, "OTU", "OTU_")
colnames(uparse_table)[-1] <- sample_names
uparse_table
write.table(uparse_table, file = file.path(result_path, "uparse/uparse_table.txt"), quote = FALSE, sep = "\t")

# load UPARSE representative sequences
uparse_seqs <- readDNAStringSet(file.path(result_path, "uparse/otus.fa"))
uparse_seqs <- as.character(uparse_seqs)
names(uparse_seqs) <- str_replace(names(uparse_seqs), "OTU", "OTU_")

```


```{r load UNOISE}

# load UNOISE table
unoise_table <- as.tibble(read.table(file = file.path(result_path, "unoise/zotu_table.txt"), header = TRUE, sep = "\t", 
                                     comment.char = ""))
colnames(unoise_table)[1] <- "id"
unoise_table$id <- str_replace(unoise_table$id, "OTU", "ZOTU_")
colnames(unoise_table)[-1] <- sample_names
unoise_table
write.table(unoise_table, file = file.path(result_path, "unoise/unoise_table.txt"), quote = FALSE, sep = "\t")

# load UNOISE sequences
unoise_seqs <- readDNAStringSet(file.path(result_path, "unoise/zotus.fa"))
unoise_seqs <- as.character(unoise_seqs)
names(unoise_seqs) <- sapply(names(unoise_seqs), FUN = str_replace, "OTU", "ZOTU_")

```


```{r load MED}

# load MED table
med_table <- read.table(file = file.path(result_path, "med/MATRIX-COUNT.txt"), header = TRUE, sep = "\t")
row.names(med_table) <- med_table[, 1]
med_table <- med_table[, -1]  # remove sample name column
med_table <- data.frame(t(med_table))   # samples as columns
med_table <- as.tibble(rownames_to_column(med_table, var = "id"))
med_table$id <- str_replace(med_table$id, "X", "Node_")
med_table

# load MED sequences
med_seqs <- readDNAStringSet(file.path(result_path, "med/NODE-REPRESENTATIVES.fasta"))
med_seqs <- as.character(med_seqs)
med_seqs <- sapply(med_seqs, FUN = str_replace, "-+", "")
names(med_seqs) <- paste0("Node_", sapply(str_split(names(med_seqs), "\\|"), FUN = `[`, 1))  # remove size annotation

# remove chimeras from MED sequences
# DADA2 method
# med_seqtab <- t(med_table)
# colnames(med_seqtab) <- med_seqs
# med_seqtab.no_chim <- removeBimeraDenovo(med_seqtab)
# med_table.no_bim <- med_table[med_seqs %in% colnames(med_seqtab.no_chim), ]

# UCHIME method
med_seqs.chimeras <- readDNAStringSet(file.path(result_path, "med/uchime/NODE-chimeras.fasta"))
if (length(med_seqs.chimeras) > 0){
  med_seqs.chimeras <- as.character(med_seqs.chimeras)
  names(med_seqs.chimeras) <- paste0("Node_", sapply(str_split(names(med_seqs.chimeras), ";"), FUN = `[`, 1))
  med_table <- med_table[!med_seqs %in% med_seqs.chimeras, ]
  med_table
  med_seqs <- med_seqs[!med_seqs %in% med_seqs.chimeras]
}

write.table(med_table, file = file.path(result_path, "med/med_table.txt"), quote = FALSE, sep = "\t")

```


```{r load Deblur table}

# load Deblur table
deblur_table <- as.tibble(read.table(file = file.path(result_path, "deblur/all.txt"), header = TRUE, sep = "\t", 
                                     skip = 1, comment.char = ""))
colnames(deblur_table)[1] <- "id"
deblur_seqs <- toupper(deblur_table$id)  
names(deblur_seqs) <- paste0("sOTU_", 1:length(deblur_seqs))
deblur_table$id <- names(deblur_seqs)
deblur_table <- deblur_table %>% select(id, noquote(sample_names))
deblur_table
write.table(deblur_table, file = file.path(result_path, "deblur/deblur_table.txt"), quote = FALSE, sep = "\t")

```


```{r load DADA2 table}

# load DADA2 table
dada2_table <- read.table(file = file.path(result_path, "dada2/sv_table.no_chim.txt"), header = TRUE, sep = "\t")
dada2_table <- data.frame(t(dada2_table))
dada2_table <- as.tibble(rownames_to_column(dada2_table, var = "id"))
dada2_seqs <- dada2_table$id
names(dada2_seqs) <- paste0("ASV_", 1:length(dada2_seqs))
dada2_table$id <- names(dada2_seqs)
dada2_table
write.table(dada2_table, file = file.path(result_path, "dada2/dada2_table.txt"), quote = FALSE, sep = "\t")

```


Remove singletons.
```{r remove singletons}

# remove rows with no more than a single observation in any sample
uclust_table <- uclust_table[apply(uclust_table[, sample_names], 1, max) > 1, ]
uparse_table <- uparse_table[apply(uparse_table[, sample_names], 1, max) > 1, ]
unoise_table <- unoise_table[apply(unoise_table[, sample_names], 1, max) > 1, ]
med_table <- med_table[apply(med_table[, sample_names], 1, max) > 1, ]
deblur_table <- deblur_table[apply(deblur_table[, sample_names], 1, max) > 1, ]
dada2_table <- dada2_table[apply(dada2_table[, sample_names], 1, max) > 1, ]

```

Add sequences to OTU tables
```{r add sequences}

uclust_table$sequence <- uclust_seqs[uclust_table$id]
uparse_table$sequence <- uparse_seqs[uparse_table$id]
unoise_table$sequence <- unoise_seqs[unoise_table$id]
med_table$sequence <- med_seqs[med_table$id]
deblur_table$sequence <- deblur_seqs[deblur_table$id]
dada2_table$sequence <- dada2_seqs[dada2_table$id]

# remove singleton sequences from sequence vectors
uclust_seqs <- uclust_seqs[uclust_table$id]
uparse_seqs <- uparse_seqs[uparse_table$id]
unoise_seqs <- unoise_seqs[unoise_table$id]
med_seqs <- med_seqs[med_table$id]
deblur_seqs <- deblur_seqs[deblur_table$id]
dada2_seqs <- dada2_seqs[dada2_table$id]

```

Read in reference sequences, and compute Levenshtein ("Hamming") distances between all inferred sequences and all reference sequences.
```{r Levenshtein function}

ref_path <- "~/projects/thesis/references"
zymo_path <- file.path(ref_path, "zymo_SSU")
ref_fastas <- list.files(zymo_path, pattern = "_16S.fasta$")

ref_seqs <- DNAStringSet()
for (f in ref_fastas){
  ref_seqs <- append(ref_seqs, readDNAStringSet(file.path(zymo_path, f)))
}
ref_seqs <- as.character(ref_seqs)
species <- sapply(str_split(names(ref_seqs), "_16S"), `[`, 1)

# define a function to compute Levenshtein distances between two sequences
levDist <- Vectorize(function(query, ref, ...) {
  mmi <- dada2:::nweval(query, ref, ...)    # this gives matches, mismatches, and indels
  ldist <- mmi[2] + mmi[3]
  if (nchar(query) > sum(mmi)) {    # query not fully overlapped by reference
    ldist <- nchar(query) - mmi[1]  # include non-overlapped nucleotides in distance
  }
  return(ldist)
})

```


Compute distances of inferred sequences to the reference sequences for each method.
```{r compute distances}

uclust_to_ref <- outer(uclust_seqs, ref_seqs, levDist, band = -1)
# Collapse distances for strains of the same species (using minimum distance)
uclust_to_ref_species <- t(aggregate(t(uclust_to_ref), list(species), min))
colnames(uclust_to_ref_species) <- uclust_to_ref_species[1, ]
uclust_to_ref_species <- uclust_to_ref_species[-1, ]
class(uclust_to_ref_species) <- "integer"

uparse_to_ref <- outer(uparse_seqs, ref_seqs, levDist, band = -1)
uparse_to_ref_species <- t(aggregate(t(uparse_to_ref), list(species), min))
colnames(uparse_to_ref_species) <- uparse_to_ref_species[1, ]
uparse_to_ref_species <- uparse_to_ref_species[-1, ]
class(uparse_to_ref_species) <- "integer"

unoise_to_ref <- outer(unoise_seqs, ref_seqs, levDist, band = -1)
unoise_to_ref_species <- t(aggregate(t(unoise_to_ref), list(species), min))
colnames(unoise_to_ref_species) <- unoise_to_ref_species[1, ]
unoise_to_ref_species <- unoise_to_ref_species[-1, ]
class(unoise_to_ref_species) <- "integer"

med_to_ref <- outer(med_seqs, ref_seqs, levDist, band = -1)
med_to_ref_species <- t(aggregate(t(med_to_ref), list(species), min))
colnames(med_to_ref_species) <- med_to_ref_species[1, ]
med_to_ref_species <- med_to_ref_species[-1, ]
class(med_to_ref_species) <- "integer"

deblur_to_ref <- outer(deblur_seqs, ref_seqs, levDist, band = -1)
deblur_to_ref_species <- t(aggregate(t(deblur_to_ref), list(species), min))
colnames(deblur_to_ref_species) <- deblur_to_ref_species[1, ]
deblur_to_ref_species <- deblur_to_ref_species[-1, ]
class(deblur_to_ref_species) <- "integer"

dada2_to_ref <- outer(dada2_seqs, ref_seqs, levDist, band = -1)
dada2_to_ref_species <- t(aggregate(t(dada2_to_ref), list(species), min))
colnames(dada2_to_ref_species) <- dada2_to_ref_species[1, ]
dada2_to_ref_species <- dada2_to_ref_species[-1, ]
class(dada2_to_ref_species) <- "integer"

```


Write a function to count and identify sequences that differ by only 1 (or 2 or 3) necleotides from a more abundant reference sequence.
```{r noisy functions}

# function to generate a table of "noisy"" sequences: those that are very similar to the reference sequences
countNoisySeqs <- function(seq_tab, dist_mat, min_dist = 1, max_dist = 3) {
  if (max(colSums(dist_mat == 0)) > 1) {
    print("WARNING: Your distance matrix contains seqeunces that are identical except for length differences.")
    print("This may result in inaccurate counts of noisy seuqences.")
  }
  
  if (sum(!sapply(seq_tab, class) %in% c("integer", "numeric")) > 0){
    stop("The sequence table must only contain column vectors of type 'numeric' or 'integer'.")
  }
  
  
  
  rmat <- matrix(0, nrow = ncol(seq_tab), ncol = (max_dist - min_dist) + 1, 
                 dimnames = list(colnames(seq_tab), paste(min_dist:max_dist, "off")))
  
  for (n in min_dist:max_dist){
    mins <- apply(dist_mat, 1, min)   
    noisy_ind <- which(dist_mat == n, arr.ind = TRUE)
    noisy_ind <- noisy_ind[dist_mat[noisy_ind] == mins[noisy_ind[, "row"]], , drop = FALSE]   # make sure it's a minimum distance
    has_zero <- apply(dist_mat[, noisy_ind[, "col"], drop = FALSE], 2, min) == 0  # is there a hit to the nearest reference?
    noisy_ind <- noisy_ind[has_zero, , drop = FALSE]
    
    if (length(noisy_ind) > 0){
      ref_row <- apply(dist_mat[, noisy_ind[, "col"] , drop = FALSE] == 0, 2, which)  # get nearest sequence that matches reference
      noisy_ind <- cbind(noisy_ind, ref_row)
      is_n_off <- ( (seq_tab[noisy_ind[, "row"], ] > 0) & 
                   (seq_tab[noisy_ind[, "row"], ] < seq_tab[noisy_ind[, "ref_row"], ]) ) * 1  # is abundance less than reference?
      row.names(is_n_off) <- row.names(noisy_ind)
      is_n_off <- aggregate(is_n_off, by = list(row.names(is_n_off)), FUN = max)
      row.names(is_n_off) <- is_n_off$Group.1
      is_n_off <- is_n_off[, -1]
      rmat[, n] <- colSums(is_n_off)
    }
    else rmat[, n] <- integer(nrow(rmat))
  }
  rmat <- cbind(rmat, "all noisy" =  rowSums(rmat))
  return(rmat)
}

# define a function that returns whether sequences of a sequence table are "noisy"
isNoisy <- function(seq_tab, dist_mat, min_dist = 1, max_dist = 3){
  if (max(colSums(dist_mat == 0)) > 1) {
    print("WARNING: Your distance matrix contains seqeunces that are identical except for length differences.")
    print("This may result in inaccurate counts of noisy seuqences.")
  }
  
  if (sum(!sapply(seq_tab, class) %in% c("integer", "numeric")) > 0){
    stop("The sequence table must only contain column vectors of type 'integer'.")
  }
  
  mins <- apply(dist_mat, 1, min)   # get minimum distance for each sequence
  noisy_ind <- which(dist_mat >= min_dist & dist_mat <= max_dist, arr.ind = TRUE) 
  if (length(noisy_ind) == 0){
    return(rep(FALSE, nrow(dist_mat)))
  } 
  else {
    noisy_ind <- noisy_ind[dist_mat[noisy_ind] == mins[noisy_ind[, "row"]], , drop = FALSE]   # make sure it's a minimum distance
    has_zero <- apply(dist_mat[, noisy_ind[, "col"], drop = FALSE], 2, min) == 0  # is there a hit to the nearest reference?
    noisy_ind <- noisy_ind[has_zero, , drop = FALSE]
    ref_row <- apply(dist_mat[, noisy_ind[, "col"] , drop = FALSE] == 0, 2, which)   # find the nearest sequence that matches the reference
    noisy_ind <- cbind(noisy_ind, ref = ref_row)
    is_noisy <- ( (seq_tab[noisy_ind[, "row"], ] > 0) & 
                    (seq_tab[noisy_ind[, "row"], ] < seq_tab[noisy_ind[, "ref"], ]) )   # is abundance less than reference?
    is_noisy <- apply(is_noisy, 1, any)   # collapse to vector
    noisy_ind <- noisy_ind[is_noisy, , drop = FALSE]    # remove any with abundance higher than reference
    noisy <- 1:nrow(dist_mat) %in% noisy_ind[, "row"]  
    return(noisy)
  }
}

```


Annotate sequences in sequence tables as "noisy" or not.
```{r annotate noisy}

uclust_table$dist_to_ref <- apply(uclust_to_ref, 1, min)
uparse_table$dist_to_ref <- apply(uparse_to_ref, 1, min)
unoise_table$dist_to_ref <- apply(unoise_to_ref, 1, min)
med_table$dist_to_ref <- apply(med_to_ref, 1, min)
deblur_table$dist_to_ref <- apply(deblur_to_ref, 1, min)
dada2_table$dist_to_ref <- apply(dada2_to_ref, 1, min)

uclust_table$hit <- uclust_table$dist_to_ref == 0
uparse_table$hit <- uparse_table$dist_to_ref == 0
unoise_table$hit <- unoise_table$dist_to_ref == 0
med_table$hit <- med_table$dist_to_ref == 0
deblur_table$hit <- deblur_table$dist_to_ref == 0
dada2_table$hit <- dada2_table$dist_to_ref == 0

uclust_table$noisy <- isNoisy(uclust_table[, sample_names], uclust_to_ref, 1, 3)
uparse_table$noisy <- isNoisy(uparse_table[, sample_names], uparse_to_ref, 1, 3)
unoise_table$noisy <- isNoisy(unoise_table[, sample_names], unoise_to_ref, 1, 3)
med_table$noisy <- isNoisy(med_table[, sample_names], med_to_ref, 1, 3)
deblur_table$noisy <- isNoisy(deblur_table[, sample_names], deblur_to_ref, 1, 3)
dada2_table$noisy <- isNoisy(dada2_table[, sample_names], dada2_to_ref, 1, 3)

uclust_table$ref_like <- uclust_table$hit | uclust_table$noisy
uparse_table$ref_like <- uparse_table$hit | uparse_table$noisy
unoise_table$ref_like <- unoise_table$hit | unoise_table$noisy
med_table$ref_like <- med_table$hit | med_table$noisy
deblur_table$ref_like <- deblur_table$hit | deblur_table$noisy
dada2_table$ref_like <- dada2_table$hit | dada2_table$noisy

```


Next, write each method's inferred (singleton-free) sequences to a fasta file, for BLASTing against nt.
```{r write fastas}

uclust_seqs <- DNAStringSet(uclust_seqs)
writeFasta(uclust_seqs, file.path(analysis_path, "uclust_seqs.fasta"))
uparse_seqs <- DNAStringSet(uparse_seqs)
writeFasta(uparse_seqs, file.path(analysis_path, "uparse_seqs.fasta"))
unoise_seqs <- DNAStringSet(unoise_seqs)
writeFasta(unoise_seqs, file.path(analysis_path, "unoise_seqs.fasta"))
med_seqs <- DNAStringSet(med_seqs)
writeFasta(med_seqs, file.path(analysis_path, "med_seqs.fasta"))
deblur_seqs <- DNAStringSet(deblur_seqs)
writeFasta(deblur_seqs, file.path(analysis_path, "deblur_seqs.fasta"))
dada2_seqs <- DNAStringSet(dada2_seqs)
writeFasta(dada2_seqs, file.path(analysis_path, "dada2_seqs.fasta"))

```


Read in the results of the BLAST searches.
```{r load blast tables}

uclust_blast <- as.tibble(read.table(file.path(analysis_path, "uclust_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
uparse_blast <- as.tibble(read.table(file.path(analysis_path, "uparse_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
unoise_blast <- as.tibble(read.table(file.path(analysis_path, "unoise_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
med_blast <- as.tibble(read.table(file.path(analysis_path, "med_blast.txt"), sep = "\t", comment.char = "#",
                                  col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
deblur_blast <- as.tibble(read.table(file.path(analysis_path, "deblur_blast.txt"), sep = "\t", comment.char = "#",
                                     col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
dada2_blast <- as.tibble(read.table(file.path(analysis_path, "dada2_blast.txt"), sep = "\t", comment.char = "#",
                                    col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))

```

Define functions to identify blast hits that are exact matches and those that are one or more nucleotides off.
```{r blast parsing functions}

isPerfectHit <- function(seq_tab, blast_file){
  blast_tab <- as.tibble(read.table(blast_file, sep = "\t", comment.char = "#",
                                    col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
  perfect_hits <- blast_tab[ (blast_tab$identity == 100) & 
                            (blast_tab$aln_length == nchar(seq_tab$sequence[match(blast_tab$seqID, seq_tab$id)])), ]
  return(seq_tab$id %in% perfect_hits$seqID)
}

isNOff <- function(seq_tab, blast_file, N = 1){
  blast_tab <- as.tibble(read.table(blast_file, sep = "\t", comment.char = "#",
                                    col.names = c("seqID", "subjectIDs", "kingdom", "identity", "aln_length", "mismatches", 
                                                  "gapopens", "gaps", "qstart", "qend", "sstart", "send", "evalue", "bitscore")))
  # find full-length hits with one mismatch or indel
  fulls <- blast_tab[ blast_tab$aln_length == nchar(seq_tab$sequence[match(blast_tab$seqID, seq_tab$id)]), ]
  one_offs <- fulls[ (fulls$mismatches == 1 & fulls$gaps == 0) | (fulls$mismatches == 0 & fulls$gaps == 1), ]   #alternate code
  
  # now check length-1 hits with 100% identity
  partials <- blast_tab[ blast_tab$aln_length == (nchar(seq_tab$sequence[match(blast_tab$seqID, seq_tab$id)]) - N), ]
  one_offs <- rbind(one_offs, partials[partials$identity == 100, ])
  is_one_off <- seq_tab$id %in% one_offs$seqID
  
  hits <- isPerfectHit(seq_tab, blast_file)
  is_one_off[hits] <- FALSE   # remove queries that also have perfect hits
  
  return(is_one_off)
}

```


Annotate sequence tables with results from BLAST analysis.
```{r annotate blast results}

uclust_table$perfect <- isPerfectHit(uclust_table, file.path(analysis_path, "uclust_blast.txt")) & !uclust_table$hit
uclust_table$one_off <- isNOff(uclust_table, file.path(analysis_path, "uclust_blast.txt"), 1)
uclust_table <- uclust_table %>% mutate(contaminant = perfect | one_off, other = !ref_like & !contaminant)

uparse_table$perfect <- isPerfectHit(uparse_table, file.path(analysis_path, "uparse_blast.txt")) & !uparse_table$hit
uparse_table$one_off <- isNOff(uparse_table, file.path(analysis_path, "uparse_blast.txt"), 1)
uparse_table <- uparse_table %>% mutate(contaminant = perfect | one_off, other = !ref_like & !contaminant)

unoise_table$perfect <- isPerfectHit(unoise_table, file.path(analysis_path, "unoise_blast.txt")) & !unoise_table$hit
unoise_table$one_off <- isNOff(unoise_table, file.path(analysis_path, "unoise_blast.txt"), 1)
unoise_table <- unoise_table %>% mutate(contaminant = perfect | one_off, other = !ref_like & !contaminant)

med_table$perfect <- isPerfectHit(med_table, file.path(analysis_path, "med_blast.txt")) & !med_table$hit
med_table$one_off <- isNOff(med_table, file.path(analysis_path, "med_blast.txt"), 1)
med_table <- med_table %>% mutate(contaminant = perfect | one_off, other = !ref_like & !contaminant)

deblur_table$perfect <- isPerfectHit(deblur_table, file.path(analysis_path, "deblur_blast.txt")) & !deblur_table$hit
deblur_table$one_off <- isNOff(deblur_table, file.path(analysis_path, "deblur_blast.txt"), 1)
deblur_table <- deblur_table %>% mutate(contaminant = perfect | one_off, other = !ref_like & !contaminant)

dada2_table$perfect <- isPerfectHit(dada2_table, file.path(analysis_path, "dada2_blast.txt")) & !dada2_table$hit
dada2_table$one_off <- isNOff(dada2_table, file.path(analysis_path, "dada2_blast.txt"), 1)
dada2_table <- dada2_table %>% mutate(contaminant = perfect | one_off, other = !ref_like & !contaminant)

# Annotate "consensus" sequences-- that are observed in every sample
uclust_table$consensus <- apply(uclust_table[, sample_names], 1, min) > 0
uparse_table$consensus <- apply(uparse_table[, sample_names], 1, min) > 0
unoise_table$consensus <- apply(unoise_table[, sample_names], 1, min) > 0
med_table$consensus <- apply(med_table[, sample_names], 1, min) > 0
deblur_table$consensus <- apply(deblur_table[, sample_names], 1, min) > 0
dada2_table$consensus <- apply(dada2_table[, sample_names], 1, min) > 0

```


Compute summary tables for each method.
```{r compute method summaries}

methods <- c("uclust", "uparse", "unoise", "med", "deblur", "dada2")
method_tables <- list(uclust_table, uparse_table, unoise_table, med_table, deblur_table, dada2_table)
names(method_tables) <- methods
dist_tables <- list(uclust_to_ref, uparse_to_ref, unoise_to_ref, med_to_ref, deblur_to_ref, dada2_to_ref)
names(dist_tables) <- methods
dist_species_tables <- list(uclust_to_ref_species, uparse_to_ref_species, unoise_to_ref_species, 
                            med_to_ref_species, deblur_to_ref_species, dada2_to_ref_species)
names(dist_species_tables) <- methods

method_summary <- list()

for (i in seq_along(methods)){
  # Compute summary statistics
  method_hits <- colSums(method_tables[[i]]$hit & method_tables[[i]][, sample_names] > 0)
  method_species <- integer()
  for (s in sample_names){
    s_species <- sum(apply(dist_species_tables[[i]][method_tables[[i]][, s] > 0, ], 2, min) == 0)
    method_species <- c(method_species, s_species)
  }
  method_counts <- colSums(method_tables[[i]][, sample_names] > 0)   # total number of sequences
  method_noisy <- countNoisySeqs(method_tables[[i]][, sample_names], dist_tables[[i]], 1, 3)   # "noisy" sequences
  method_perfect <- colSums(method_tables[[i]]$perfect & method_tables[[i]][, sample_names] > 0)  # perfect BLAST hits
  method_one_off <- colSums(method_tables[[i]]$one_off & method_tables[[i]][, sample_names] > 0)  # one-off
  method_contaminant <- method_perfect + method_one_off
  method_other <- colSums(method_tables[[i]]$dist_to_ref > 3 & !method_tables[[i]]$perfect & 
                            !method_tables[[i]]$one_off & method_tables[[i]][, sample_names] > 0)   # non-hits, not "noisy"
  method_pct_hit <- colSums(method_tables[[i]][method_tables[[i]]$hit, sample_names]) / colSums(method_tables[[i]][, sample_names])
  method_pct_noisy <- colSums(method_tables[[i]][method_tables[[i]]$noisy, sample_names]) / colSums(method_tables[[i]][, sample_names])
  method_pct_cont <- colSums(method_tables[[i]][method_tables[[i]]$contaminant, sample_names]) / colSums(method_tables[[i]][, sample_names])
  method_pct_other <- colSums(method_tables[[i]][method_tables[[i]]$other, sample_names]) / colSums(method_tables[[i]][, sample_names])
  m_summary <- data.frame(sample = sample_names, total = method_counts, hits = method_hits, species = method_species,
                          method_noisy, perfect = method_perfect, one_off = method_one_off, contaminants = method_contaminant,
                          other = method_other, 
                          pct_hit = method_pct_hit, pct_noisy = method_pct_noisy, 
                          pct_cont = method_pct_cont, pct_other = method_pct_other,
                          row.names = NULL, check.names = FALSE)
  
  method_summary[[i]] <- m_summary
}

names(method_summary) <- methods

all_summary <- do.call("rbind", method_summary) %>% as.tibble %>% mutate(method = rep(methods, each = length(sample_names)))

```


Now, compute summary tables by sample (for all 9 dilutions).
```{r compute sample tables}

# combine sequences from all methods
all_seqs <- unique(c(uclust_table$sequence, uparse_table$sequence, unoise_table$sequence, med_table$sequence, 
                     deblur_table$sequence, dada2_table$sequence))

dilution_table <- list()
dilution_to_ref <- list()
dilution_summary <- list()

for (i in seq_along(sample_names)){
  # create a sequence table for the sample
  sample_st <- matrix(as.integer(0), ncol = length(all_seqs), nrow = length(methods))
  colnames(sample_st) <- all_seqs
  rownames(sample_st) <- methods
  
  # populate the table with sample abundances
  sample_st["uclust", uclust_table$sequence] <- uclust_table[[ sample_names[i] ]]
  sample_st["uparse", uparse_table$sequence] <- uparse_table[[ sample_names[i] ]]
  sample_st["unoise", unoise_table$sequence] <- unoise_table[[ sample_names[i] ]]
  sample_st["med", med_table$sequence] <- med_table[[ sample_names[i] ]]
  sample_st["deblur", deblur_table$sequence] <- deblur_table[[ sample_names[i] ]]
  sample_st["dada2", dada2_table$sequence] <- dada2_table[[ sample_names[i] ]]

  # remove columns (sequences) with all zeros
  sample_st <- sample_st[, colSums(sample_st) > 0]
  
  # collapse sequences that only differ in length
  sample_st <- collapseNoMismatch(sample_st)
  
  # Compute distances from sample sequences to reference
  sample_to_ref <- outer(colnames(sample_st), ref_seqs, levDist, band = -1)
  row.names(sample_to_ref) <- colnames(sample_st)
  # Collapse distances for strains of the same species (using minimum distance)
  sample_to_ref_species <- t(aggregate(t(sample_to_ref), list(species), min))
  colnames(sample_to_ref_species) <- sample_to_ref_species[1, ]
  sample_to_ref_species <- sample_to_ref_species[-1, ]
  class(sample_to_ref_species) <- "integer"

  # Convert to data.frame and annotate sequences
  sample_table <- data.frame(t(sample_st)) %>% rownames_to_column(var = "sequence") %>% as.tibble()
  sample_table$dist_to_ref <- apply(sample_to_ref, 1, min)
  sample_table$hit <- sample_table$dist_to_ref == 0
  sample_table$noisy <- isNoisy(sample_table[, methods], sample_to_ref)
  
  # Compute summary statistics
  sample_hits <- colSums(sample_table$hit & sample_table[, methods] > 0)
  sample_species <- integer()
  for (m in methods){
    m_species <- sum(apply(sample_to_ref_species[sample_table[[m]] > 0, ], 1, min) == 0)
    sample_species <- c(sample_species, m_species)
  }
  sample_counts <- colSums(sample_table[, methods] > 0)   # total number of sequences
  sample_pct_hit <- colSums(sample_table[sample_table$hit, methods]) / colSums(sample_table[, methods])
  #sample_table[, methods] <- sapply(sample_table[, methods], as.integer)
  sample_noisy <- countNoisySeqs(sample_table[, methods], sample_to_ref, 1, 3)   # "noisy" sequences
  sample_other <- colSums(sample_table$dist_to_ref > 3 & sample_table[, methods] > 0)   # non-hits, not "noisy"
  sample_summary <- data.frame(method = methods, total = sample_counts, hits = sample_hits, species = sample_species,
                             sample_noisy, other = sample_other, hit_rate = sample_pct_hit,
                             row.names = NULL, check.names = FALSE)
  
  dilution_table[[i]] <- sample_table
  dilution_to_ref[[i]] <- sample_to_ref
  dilution_summary[[i]] <- sample_summary
  cat("Done with sample", sample_names[i], "\n")
}

names(dilution_table) <- sample_names
names(dilution_to_ref) <- sample_names
names(dilution_summary) <- sample_names

```



Reformat annotated sequence tables for plotting with `ggplot`.
```{r prep for ggplot}

uclust_gg <- gather(uclust_table, "sample", "count", one_of(sample_names))
median_uclust <- median(colSums(uclust_table[, sample_names]))
uclust_gg <- uclust_gg %>% mutate(log10 = log10(count))
uclust_gg <- uclust_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_uclust), 
         log10_norm = log10(norm_median))
uclust_gg <- uclust_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

uparse_gg <- gather(uparse_table, "sample", "count", one_of(sample_names))
median_uparse <- median(colSums(uparse_table[, sample_names]))
uparse_gg <- uparse_gg %>% mutate(log10 = log10(count))
uparse_gg <- uparse_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_uparse), 
         log10_norm = log10(norm_median))
uparse_gg <- uparse_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

unoise_gg <- gather(unoise_table, "sample", "count", one_of(sample_names))
median_unoise <- median(colSums(unoise_table[, sample_names]))
unoise_gg <- unoise_gg %>% mutate(log10 = log10(count))
unoise_gg <- unoise_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_unoise), 
         log10_norm = log10(norm_median))
unoise_gg <- unoise_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

med_gg <- gather(med_table, "sample", "count", one_of(sample_names))
median_med <- median(colSums(med_table[, sample_names]))
med_gg <- med_gg %>% mutate(log10 = log10(count))
med_gg <- med_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_med), 
         log10_norm = log10(norm_median))
med_gg <- med_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

deblur_gg <- gather(deblur_table, "sample", "count", one_of(sample_names))
median_deblur <- median(colSums(deblur_table[, sample_names]))
deblur_gg <- deblur_gg %>% mutate(log10 = log10(count))
deblur_gg <- deblur_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_deblur), 
         log10_norm = log10(norm_median))
deblur_gg <- deblur_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

dada2_gg <- gather(dada2_table, "sample", "count", one_of(sample_names))
median_dada2 <- median(colSums(dada2_table[, sample_names]))
dada2_gg <- dada2_gg %>% mutate(log10 = log10(count))
dada2_gg <- dada2_gg %>% group_by(sample) %>%
  mutate(norm = count / sum(count), 
         norm_median = round(norm * median_dada2), 
         log10_norm = log10(norm_median))
dada2_gg <- dada2_gg %>% select(id, sample, count, everything()) %>% select(-sequence, sequence)

all_table <- do.call("rbind", method_tables) %>% mutate(method = rep(methods, times = sapply(method_tables, nrow)))
all_table <- gather(all_table, "sample", "count", one_of(sample_names)) %>% select(-sequence, id, sample, count, everything())

medians <- all_table %>% group_by(method, sample) %>% summarise(reads = sum(count)) %>% summarise(median_reads = median(reads)) %>%
  select(median_reads)

dilution_labels <- sapply(str_split(sample_names, "\\."), function(d){
  paste(d[3:4], collapse = ":")
})
dilution_labels[1] <- "1:1"
dilution_values <- sapply(str_split(dilution_labels, ":"), function(d) as.numeric(d[1]) / as.numeric(d[2]))

```


Create tables of sequence and read count summaries per sample, and for each method.
```{r summary tables}

load(file.path(result_path, "dada2/tables.RData"))
median_raw <- median(track_table[, "raw"])

sample_seq_count <- sapply(method_tables, function(st, ss) colSums(st[, ss] > 0), sample_names) %>%
  t() %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()
sample_seq_count_gg <- gather(sample_seq_count, "sample", "seqs", -method)

sample_read_count <- sapply(method_tables, function(st, ss) colSums(st[, ss]), sample_names) %>%
  t() %>% as.data.frame() %>% rownames_to_column(var = "method") %>% as.tibble()
sample_read_count_gg <- gather(sample_read_count, "sample", "reads", -method)

sample_summary <- merge(sample_read_count_gg, sample_seq_count_gg, by = c("method", "sample"))
sample_summary <- sample_summary %>% group_by(method) %>% 
  mutate(seqs_norm_method = round(seqs * median(reads) / reads),
         seqs_norm_raw = round(seqs * median_raw / reads))

# Calculate mean signal strengths of each method's reference and non-reference sequences
# sample_sn <- lapply(method_tables, function(m){
#   sn_mean <- m[, sample_names] %>% sapply(., tapply, m$hit, function(x) mean(log10(x[x > 0])))
#   sn_mean <- as.tibble(rownames_to_column(as.data.frame(sn_mean), var = "hit"))
#   return(sn_means)
# })
sample_sn <- lapply(method_tables, function(m){
  sn_mean <- m[, sample_names] %>% sapply(., tapply, m$hit, function(x) mean(log10(x[x > 0])))
  sn_mean <- sn_mean["TRUE", ] / sn_mean["FALSE", ]
  #sn_mean <- as.tibble(rownames_to_column(as.data.frame(sn_mean), var = "hit"))
  return(sn_mean)
})
sample_sn <- do.call("rbind", sample_sn) %>% as.data.frame %>% rownames_to_column(var = "method") %>% as.tibble
#sample_sn$hit <- as.logical(sample_sn$hit)
#sample_sn$method <- rep(methods, each = 2)
sample_sn_gg <- gather(sample_sn, "sample", "mean_snr", one_of(sample_names))
sample_sn_gg 

sample_summary <- merge(sample_summary, sample_sn_gg, by = c("method", "sample")) %>% as.tibble()

```


Compute Bray-Curtis similarity between each dilution sample and the neat sample, for each method.
```{r Bray-Curtis}

# normalize sequence table by median read count per method
norm_median_tables <- lapply(method_tables, function(table){
  seq_tab <- table[table$hit, c("id", sample_names)]
  medn <- median(colSums(seq_tab[, sample_names]))
  seq_tab[, sample_names] <- sapply(seq_tab[, sample_names], function(x) round(medn * (x / sum(x))))
  seq_tab <- as.tibble(seq_tab)
  return(seq_tab)
})

# compute Bray-Curtis similarity relative to neat sample
bray_curtis <- lapply(norm_median_tables, function(table){
  neat <- table[[sample_names[1]]]
  bc <- sapply(table[, sample_names], function(sample){
    return(2 * sum(pmin(neat, sample)) / (sum(neat) + sum(sample)))
  })
  return(bc)
})

bray_curtis <- do.call("rbind", bray_curtis) %>% as.tibble() %>% mutate(method = methods) %>% select(method, everything())

```


Investigate how many of the filtered input reads are represented in the processed sequence tables. Is there much of a difference in how many reads are "kept" at different dilutions, or by different methods?
```{r reads kept}

data_path <- "~/projects/thesis/data/dilution"
prefixes <- unique(sapply(str_subset(list.files(file.path(data_path, "raw")), "s\\d+.*"), str_replace, "_R.*", ""))
raw_reads <- sapply(prefixes, function(f) length(readFastq(file.path(data_path, "raw"), pattern = paste0(f, "_R1"))))
filtered_reads <- sapply(prefixes, function(f) length(readFastq(file.path(data_path, "filtered"), pattern = f)))

kept_reads <- matrix(0, nrow = length(methods), ncol = length(sample_names))
row.names(kept_reads) <- methods
colnames(kept_reads) <- sample_names

kept_reads["uclust", ] <- colSums(uclust_table[, sample_names]) / filtered_reads
kept_reads["uparse", ] <- colSums(uparse_table[, sample_names]) / raw_reads
kept_reads["unoise", ] <- colSums(unoise_table[, sample_names]) / raw_reads
kept_reads["med", ] <- colSums(med_table[, sample_names]) / filtered_reads
kept_reads["deblur", ] <- colSums(deblur_table[, sample_names]) / filtered_reads
kept_reads["dada2", ] <- track_table[, "non_chim"] / track_table[, "raw"]

kept_reads <- as.tibble(rownames_to_column(as.data.frame(kept_reads), var = "method"))
kept_reads_gg <- gather(kept_reads, "sample", "pct_kept", sample_names)

library(directlabels)

# plot kept reads across dilutions, color-coded by method
kept_lines <- ggplot(kept_reads_gg, aes(x = sample, y = pct_kept)) +
  geom_line(aes(group = method, color = method)) +
  geom_dl(aes(label = method), method = list(dl.combine("first.qp", "last.qp"))) +
  labs(title = "Fraction of input reads represented in the processed output",
          x = "sample dilution", y = "proportion of input reads") + 
  scale_x_discrete(labels = dilution_labels)
kept_lines

```


Examine the relationship between read count and dilution at the sequence level. Are they correlated, and is the correlation positive or negative?
```{r dilution correlation}

seqCor <- function(seq_tab, samples, vals) {
  sapply(seq_along(seq_tab[[1]]), function(row) {
    cor(seq_tab %>% select(samples) %>% slice(row) %>% unlist(., use.names = FALSE), vals, method = "spearman")
  })
}

method_tables <- lapply(method_tables, function(mt) {
  mt$dilution_cor <- seqCor(mt, sample_names, dilution_values)
  return(mt)
})

# What are the dilution correlations for the reference sequences?
method_tables$uclust %>% filter(hit) %>% select(sample_names, dilution_cor)

# How many sequences have very strong negative correlations?
method_tables$uclust %>% filter(dilution_cor <= -.9, contaminant) %>% select(sample_names, dilution_cor)

# How many sequences are "contaminants" and have less than three zero sequence counts?
method_tables$dada2 %>% filter(contaminant, !noisy, rowSums(.[, sample_names] == 0) < 3) %>% select(sample_names, dilution_cor)

# Of these, how many have negative dilution correlations?
method_tables$dada2 %>% filter(contaminant, !noisy, rowSums(.[, sample_names] == 0) < 3, dilution_cor < 0) %>% 
  select(sample_names, dilution_cor)

# How many unclassified ("other") sequences have less than three zero counts?
method_tables$dada2 %>% filter(other, !noisy, rowSums(.[, sample_names] == 0) < 4) %>% select(sample_names, dilution_cor)

# Of these, how many have negative dilution correlations?
method_tables$uparse %>% filter(other, !noisy, rowSums(.[, sample_names] == 0) < 4, dilution_cor < 0) %>% 
  select(sample_names, dilution_cor)

```


```{r read count histograms}

library(ggridges)

all_table_pos <- all_table %>% filter(sample == sample_names[9], count > 100)

read_hist <- ggplot(all_table_pos, aes(x = count, y = hit)) +
  geom_density_ridges(aes(fill = hit, height = ..density..), scale = 0.9, stat = "binline", bins = 120, panel_scaling = FALSE) +
  geom_vline(data = all_table_pos %>% filter(hit) %>% group_by(method) %>% summarise(min = min(count)), aes(xintercept = min)) +
  facet_wrap(~method) +
  scale_fill_brewer(palette = "Dark2")
read_hist

```


Create plots to show relative strength of signal (reference hits) to noise (non-reference hits).
```{r snr plots}

all_table <- all_table %>% group_by(method, sample) %>% 
  mutate(tot_reads = sum(count), norm = count / tot_reads) %>% 
  ungroup() %>% group_by(method) %>%
  mutate(norm_median = round(norm * median(tot_reads)), log10_norm = log10(norm_median)) %>% 
  ungroup() %>%
  select(-sequence, -tot_reads, sequence)

theme_set(theme_bw())

meth = "deblur"

# ref vs. non-ref boxplots, raw counts
snr_box <- ggplot(data = all_table %>% filter(method == meth)) +
  geom_boxplot(aes(x = sample, y = log10_norm, fill = factor(hit, levels = c(TRUE, FALSE)))) +
  labs(title = "Distribution of log10(read counts) for reference and non-reference sequences",
          subtitle = "Dilution series processed with DADA2", x = "sample dilution", y = "log10(read counts)") +
  scale_fill_manual(name = "sequence identity", labels = c("reference", "non-reference"), values = c("blue", "orange")) +
  scale_x_discrete(labels = dilution_labels) +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
snr_box

# ref vs. non-ref violin plots, raw counts
snr_violin <- ggplot(data = all_table %>% filter(method == method)) +
  geom_violin(aes(x = sample, y = log10_norm, fill = factor(hit, levels = c(T, F)))) +
  labs(title = "Density distribution of log10(read counts) for reference and non-reference sequences",
          subtitle = "Dilution series processed with DADA2", x = "sample dilution", y = "log10(read counts)") +
  scale_fill_manual(name = "sequence identity", labels = c("reference", "non-reference"), values = c("blue", "orange")) +
  scale_x_discrete(labels = dilution_labels) +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
snr_violin

# average (mean(log10)) of ref vs. non-ref signal
# mean_snr_line <- ggplot(sample_sn_gg, aes(x = sample, y = mean_log10)) +
#   geom_point(data = subset(sample_sn_gg, hit), aes(color = method, group = method)) +
#   geom_line(data = subset(sample_sn_gg, hit), aes(color = method, group = method)) + 
#   geom_point(data = subset(sample_sn_gg, !hit), aes(color = method, group = method)) +
#   geom_line(data = subset(sample_sn_gg, !hit), aes(color = method, group =method))
# mean_snr_line

mean_snr_line <- ggplot(sample_summary, aes(x = sample, y = mean_snr)) +
  geom_point(aes(color = method, group = method)) +
  geom_line(aes(color = method, group = method)) +
  scale_x_discrete(labels = dilution_labels) +
  scale_color_brewer(palette = "Dark2")
mean_snr_line

```

Create plot to show sample composition, colored by reference vs. non-reference.
```{r composition plot}

library(RColorBrewer)

meth = "dada2"

set.seed(444)
data_table <- method_tables[[meth]]
data_gg <- all_table %>% filter(method == meth)

hit_colors <- brewer.pal(sum(data_table$hit), "Dark2")
other_colors <- sample(grey.colors(5, start = 0.5, end = 0.9), sum(!data_table$hit), replace = TRUE)

data_table$color <- rep(NA, nrow(data_table))
data_table$color[data_table$hit] <- hit_colors
data_table$color[!data_table$hit] <- other_colors

color_gg <- data_table %>% select(id, sample_names, color) %>% gather("sample", "count", sample_names)
legend_color <- c(bright = hit_colors[2], dull = other_colors[2])
data_gg <- inner_join(data_gg, color_gg)

comp_bar <- ggplot(data_gg, aes(x = sample, y = count)) +
  geom_col(aes(fill = color, group = hit, alpha = ifelse(hit, "bright", "dull")), width = 0.5, position = position_fill()) +
  scale_fill_identity(guide = FALSE) +
  scale_alpha_manual(name = "Sequence type", 
                     labels = c("mock community", "other"),
                     values = c(bright = 1, dull = 1), 
                     guide = guide_legend(override.aes = list(fill = c(hit_colors[4], "#AEAEAE")), 
                                          keywidth = NULL, keyheight = NULL)) +
  scale_x_discrete(labels = dilution_labels) +
  labs(title = "Relative composition of dilution series samples", x = "sample dilution", y = "fraction of sample") +
  theme(legend.position = "right", legend.title = element_text(size = 12), 
        axis.text = element_text(size = 12),
        axis.title = element_text(size = 16),
        plot.title = element_text(size = 20))
comp_bar
#comp_bar + coord_polar(theta = "y", start = 0)

```

Plot number of inferred sequences for each sample and each method.
```{r sequence count plots}

# non-normalized counts from each method
seqs_line_non <- ggplot(sample_summary, aes(x = sample, y = seqs)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences/OTUs at each concentration by each method", subtitle = "Non-normalized counts",
       x = "sample dilution", y = "sequence/OTU count") +
  scale_color_brewer(palette = "Dark2") +
  scale_x_discrete(labels = dilution_labels) +
  theme(legend.position = c(0.95, 0.95),
        legend.justification = c(1, 1))
seqs_line_non

# normalized by median read count after processing by each method
seqs_line_norm <- ggplot(sample_summary, aes(x = sample, y = seqs_norm_method)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences/OTUs at each concentration by each method", 
       subtitle = "Normalized by median read counts",
       x = "sample dilution", y = "sequence/OTU count") +
  scale_color_brewer(palette = "Dark2") +
  scale_x_discrete(labels = dilution_labels) +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
seqs_line_norm

# same plot as above, but only the first seven samples
seqs_line_norm2 <- ggplot(sample_summary[sample_summary$sample %in% sample_names[1:7], ], aes(x = sample, y = seqs_norm_method)) +
  geom_point(aes(color = method)) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences/OTUs at each concentration by each method", 
       subtitle = "Normalized by median read counts (first 7 dilutions only)",
       x = "sample dilution", y = "sequence/OTU count") +
  scale_x_discrete(labels = dilution_labels) +
  scale_color_brewer(palette = "Dark2")
seqs_line_norm2

# normalized by median count of raw (un-processed) reads
seqs_line_norm_raw <- ggplot(sample_summary, aes(x = sample, y = seqs_norm_raw)) +
  geom_point(aes(color = method), size = 2) +
  geom_line(aes(group = method, color = method)) +
  labs(title = "Number of inferred sequences/OTUs at each concentration by each method", 
       subtitle = "Normalized by median raw reads",
       x = "sample dilution", y = "sequence/OTU count") +
  scale_color_brewer(palette = "Dark2") +
  scale_x_discrete(labels = dilution_labels) +
  theme(legend.position = c(0.05, 0.95),
        legend.justification = c(0, 1))
seqs_line_norm_raw

```


Plot bar plots of relative abundance of reference sequences, facetted by sample.
```{r reference seq bar plots}

ref_bars <- ggplot(data = all_table %>% filter(hit, method == "deblur")) +
  geom_col(aes(x = id, y = norm, fill = id), width = 0.8) +
  facet_wrap(~sample) +
  scale_fill_brewer(palette = "Dark2")
ref_bars

# ref_joy <- ggplot(data = all_table %>% filter(hit, method == "uclust")) +
#   geom_density_ridges(aes(x = id, y = sample, fill = id, height = ..density..), stat = "density")
# ref_joy

```


Plot read counts of reference sequences across dilutions to show whether abundances maintain their relative order.
```{r reference read counts}

ref_lines <- ggplot(data = all_table %>% filter(hit, method == "dada2"), aes(x = sample, y = norm_median)) +
  geom_line(aes(group = id, color = id)) +
  geom_point(aes(color = id)) +
  scale_x_discrete(labels = dilution_labels) +
  scale_color_brewer(palette = "Dark2")
  #facet_wrap(~method)
ref_lines

```


Plot Bray-Curtis distances of each dilution sample's reference sequences relative to the neat sample for each method.
```{r bray-curtis lines}

bc_gg <- gather(bray_curtis, "sample", "bc_similarity", one_of(sample_names))

bc_lines <- ggplot(data = bc_gg, aes(x = sample, y = bc_similarity, color = method)) +
  geom_line(aes(group = method)) +
  geom_point() +
  scale_x_discrete(labels = dilution_labels) +
  scale_color_brewer(palette = "Dark2")
bc_lines

```


Plot distribution boxplots of non-hit read counts for "contaminant" sequences vs non-contaminant ("other") sequences.
```{r snr contaminant vs other}

#snr_box_cont <- ggplot(dada2_gg[!dada2_gg$hit & !dada2_gg$noisy, ], aes(x = sample, y = log10_norm)) +
snr_box_cont <- ggplot(all_table %>% filter(method == "dada2", !ref_like), aes(x = sample, y = log10_norm)) +
  geom_boxplot(aes(fill = factor(contaminant, levels = c(TRUE, FALSE)))) +
  labs(title = "Distribution of log10(read counts) for non-hit contaminant and non-contaminant sequences",
          subtitle = "Dilution series processed with DADA2", x = "sample dilution", y = "log10(read counts)") +
  scale_fill_manual(name = "sequence identity", labels = c("contaminant", "non-contaminant"), values = c("blue", "orange")) +
  scale_x_discrete(labels = dilution_labels) +
  theme(legend.position = c(0.95, 0.05),
        legend.justification = c(1, 0))
snr_box_cont

```


Plot bar plots of summary stats (number of hits, noisies, contaminants, other).
```{r summary stat bar plots}

all_summary_gg <- gather(all_summary, "statistic", "value", -sample, -method)

hits_median <- all_table %>% filter(hit) %>% .[["count"]] %>% mean()
noisy_median <- all_table %>% filter(noisy) %>% .[["count"]] %>% mean()
contam_median <- all_table %>% filter(contaminant) %>% .[["count"]] %>% mean()
other_median <- all_table %>% filter(noisy) %>% .[["count"]] %>% mean()

all_summary_gg <- all_summary_gg %>% mutate(norm_median = rep(1, nrow(all_summary_gg)))
all_summary_gg[all_summary_gg$statistic == "hits", "norm_median"] <- all_summary_gg %>% 
  filter(statistic == "hits") %>% .[["value"]] * log10(hits_median)
all_summary_gg[all_summary_gg$statistic == "noisy", "norm_median"] <- all_summary_gg %>% 
  filter(statistic == "noisy") %>% .[["value"]] * log10(noisy_median)
all_summary_gg[all_summary_gg$statistic == "contaminants", "norm_median"] <- all_summary_gg %>% 
  filter(statistic == "contaminants") %>% .[["value"]] * log10(contam_median)
all_summary_gg[all_summary_gg$statistic == "other", "norm_median"] <- all_summary_gg %>% 
  filter(statistic == "other") %>% .[["value"]] * log10(other_median)

bar_colors <- brewer.pal(9, "Set1")

summary_bars <- ggplot(data = all_summary_gg %>% filter(statistic %in% c("hits", "all noisy", "contaminants", "other")), 
                       aes(x = sample)) +
  geom_col(aes(y = log2(1 + value), fill = statistic), position = position_dodge()) +
  facet_wrap(~method) +
  scale_fill_manual(values = hit_colors[c(1, 3, 4, 6)])
  #scale_fill_brewer(palette = "Set1")
summary_bars

summary_bars <- ggplot(data = all_summary_gg %>% filter(statistic %in% c("pct_hit", "pct_noisy", "pct_cont", "pct_other")), 
                       aes(x = sample)) +
  geom_col(aes(y = value, fill = statistic), position = position_dodge()) +
  facet_wrap(~method) +
  scale_fill_manual(values = bar_colors[c(3, 4, 5, 2)])
  #scale_fill_brewer(palette = "Set1")
summary_bars

```


Plot ratio of dada2 ASV_3 to ASV_8.
```{r}

# ratio <-  dada2_table %>% filter(id %in% c("ASV_3", "ASV_8")) %>% select(sample_names) 
# ratio <- ratio[1,] / ratio[2,]
# ratio <- gather(ratio, "sample", "ratio")
# 
# ggplot(data = dada2_gg %>% filter(id %in% c("ASV_3", "ASV_8")), aes(x = sample, y = count)) +
#   geom_line(aes(group = id, color = id)) +
#   geom_point(aes(color = id))
# 
# ggplot(ratio, aes(x = sample, y = ratio)) +
#   geom_line() +
#   geom_point()

```


Here, I'm going to perform a very simple analysis with the `phyloseq` package. The idea is to mimic a "typical" pre-processing analysis that a microbiome researcher might do with a dataset of biological sample(s). Namely, I'm going to investigate the effect of two common filtering/data reduction techniques applied to the already-clustered sequence/OTU tables:
1. Filtering candidate sequences/OTUs using a relative abundance threshold (e.g. 0.1% of total sample read count)
2. Combining similar sequences/OTUs using a taxonomy table by agglomerating to the genus level
First, I'll create `phyloseq` object to facilitate the analysis.
```{r create phyloseq}

library("phyloseq")

# extract sequence matrices of read counts only from the method sequence tibbles
seq_mats <- lapply(method_tables, function(mt){
  seq_mat <- as.matrix(mt[, sample_names])
  rownames(seq_mat) <- mt$id
  return(seq_mat)
})

# load taxonomy tables
tax_mats <- lapply(list.files(result_path), function(f){
  tax_table <- read.table(file.path(result_path, f, paste0(f, "_taxa.txt")), header = TRUE, sep = "\t")
  tax_mat <- as.matrix(tax_table)
  return(tax_mat)
})
names(tax_mats) <- list.files(result_path)
tax_mats <- tax_mats[methods]

# extract sequences from taxonomy tables and convert to DNAStringSet
seqs_list <- lapply(tax_mats, function(t) DNAStringSet(t[, 1]))

# remove sequences from taxonomy tables
tax_mats <- lapply(tax_mats, function(t) t[, -1])

# construct phyloseq OTU and taxonomy tables
otu_tabs <- lapply(seq_mats, otu_table, taxa_are_rows = TRUE)
tax_tabs <- lapply(tax_mats, tax_table)

# construct phyloseq objects
physeqs <- mapply(phyloseq, otu_tabs, tax_tabs, seqs_list, SIMPLIFY = FALSE)

# remove intermediate objects
rm(seq_mats, tax_mats, seqs_list, otu_tabs, tax_tabs)

```


Now I'll implement the filtering described above.
```{r phyloseq filter}

# convert sequence tables to relative abundances
read_counts <- lapply(physeqs, sample_sums)
physeqs.rel <- lapply(physeqs, transform_sample_counts, function(x) x / sum(x))

# remove observations that are below 0.1% relative abundance
physeqs.ab0.1 <- lapply(physeqs.rel, function(ps) {
  transform_sample_counts(ps, function(x) ifelse(x >= 1e-3, x, 0))
})
physeqs.ab0.1 <- lapply(physeqs.ab0.1, function(ps) {
  prune_taxa(taxa_sums(ps) > 0, ps)
})
invisible(mapply(function(ps, psa, id) cat("Total number of mock community sequences before filtering", toupper(id), "table:", ntaxa(ps), "\n",
                                           "Total number after filtering out sequences below 0.1% relative abundance:", ntaxa(psa), "\n\n"),
                 physeqs, physeqs.ab0.1, names(physeqs)))

# alternatively, agglomerate each OTU table to the genus level
physeqs.genus <- lapply(physeqs, tax_glom, taxrank = "Genus", NArm = FALSE)

# now apply both filters in sequence
physeqs.ab0.1.genus <- lapply(physeqs.ab0.1, tax_glom, taxrank = "Genus", NArm = FALSE)

```


Finally, convert the `phyloseq` OTU tables back to tibbles, and annotate them as was done above.
```{r annotate filtered tables}

# extract OTU tables from phyloseq objects
method_tables.filt <- mapply(function(ps, rc) {
  table <- otu_table(ps) 
  table <- sweep(table, MARGIN = 2, STATS = rc, `*`)
  table <- table %>% as.data.frame() %>% rownames_to_column(var = "id") %>% as.tibble()
  table$sequence <- refseq(ps) %>% as.character()
  return(table)
  },
  physeqs.ab0.1.genus, read_counts, SIMPLIFY = FALSE)

# annotate the filtered OTU tables
method_tables.filt <- mapply(function(mf, mt) {
  midx <- match(mf$sequence, mt$sequence)
  vars <- c("id", "dist_to_ref", "hit", "noisy", "ref_like", "perfect", "one_off", "contaminant", "other", "consensus", "dilution_cor")
  ann_tab <- mt[midx, vars]
  ann_tab <- full_join(mf, ann_tab, by = "id")
  return(ann_tab)
}, method_tables.filt, method_tables, SIMPLIFY = FALSE)

```

